{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14903\n",
      "4816\n",
      "464\n"
     ]
    }
   ],
   "source": [
    "#barcode redundancy theoretical calculator (stolen fro Nat, though this is really just the birthday problem)\n",
    "import random\n",
    "def draw(barcodes, draws): \n",
    "    barcode_list = list(range(barcodes)) \n",
    "    duplicates = 0 \n",
    "    barcodes_found = set() \n",
    "    for i in range(draws): \n",
    "        barcode = random.choice(barcode_list) \n",
    "        if barcode in barcodes_found: \n",
    "            duplicates += 1 \n",
    "        barcodes_found.add(barcode) \n",
    "    return duplicates\n",
    "\n",
    "print(draw(300000,100000))\n",
    "# The mutants and then pick 100000 barcodes after this is how many barcodes will have 2 mutants)\n",
    "print(draw(1000000,100000))\n",
    "print(draw(10000000,100000))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n"
     ]
    }
   ],
   "source": [
    "#Here I parse sequence file and print out the number of reads in the file\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "mockDataParsed = SeqIO.parse(\"NP_11_18_13_mockData.txt\", \"fastq\")\n",
    "\n",
    "readCounter = 0\n",
    "\n",
    "for record in mockDataParsed:\n",
    "    readCounter += 1\n",
    "\n",
    "print(readCounter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDQSSRYVNLALKEEDLIAGGEHVLCAYIMKPKAGYGYVATAAHFAAESSTGTNVEVCTTDDFTRGVDALVYEVDEARELTKIAYPVALFDRNITDGKAMIASFLTLTMGNNQGMGDVEYAKMHDFYVPEAYRALFDGPSVNISALWKVLGRPEVDGGLVVGTIIKPKLGLRPKPFAEACHAFWLGGDFIKNDEPQGNQPFAPLRDTIALVADAMRRAQDETGEAKLFSANITADDPFEIIARGEYVLETFGENASHVALLVDGYVAGAAAITTARRRFPDNFLHYHRAGHGAVTSPQSKRGYTAFVHCKMARLQGASGIHTGTMGFGKMEGESSDRAIAYMLTQDEAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTAGGGAFGHIDGPVAGARSLRQAWQAWRDGVPVLDYAREHKELARAFESFPGDADQIYPGWRKALGVEDTRSALPA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/NP_11_dataAnalysis/lib/python3.7/site-packages/Bio/Seq.py:2742: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "#this is the template for alignment and should be a fasta file with only one sequence in it\n",
    "templateParsed = SeqIO.parse('rbcL_codonopt_barcoded.txt', 'fasta')\n",
    "TEMPLATE_STRING = None\n",
    "for record in templateParsed:\n",
    "    assert TEMPLATE_STRING == None\n",
    "    TEMPLATE_STRING = str(record.seq)\n",
    "    RCtemplateString = record.seq.reverse_complement()\n",
    "    TEMPLATE_TRANSLATION = str(record.seq.translate(to_stop=True))\n",
    "\n",
    "SEARCH_MOTIF_LENGTH = 11\n",
    "\n",
    "# This is the FULL sequence between the barcode and the last coding nucleotide (including stop codons)\n",
    "BARCODE_FLANK = 'CCTGCAGGCTATTA'\n",
    "\n",
    "print(TEMPLATE_TRANSLATION)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/NP_11_dataAnalysis/lib/python3.7/site-packages/Bio/Seq.py:2742: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Read ID</th>\n",
       "      <th>Sequence1</th>\n",
       "      <th>Sequence2</th>\n",
       "      <th>translation1</th>\n",
       "      <th>translation2</th>\n",
       "      <th>Average phredQ1</th>\n",
       "      <th>Average phredQ2</th>\n",
       "      <th>Number of bases with phredQ under 20 read 1</th>\n",
       "      <th>Number of bases with phredQ under 20 read 2</th>\n",
       "      <th>Is there a barcode in read 1?</th>\n",
       "      <th>Is there a barcode in read 2?</th>\n",
       "      <th>Nonbarcode read start position1</th>\n",
       "      <th>Nonbarcode read start position2</th>\n",
       "      <th>Barcode sequence1</th>\n",
       "      <th>Barcode sequence2</th>\n",
       "      <th>Mutation1</th>\n",
       "      <th>Mutation2</th>\n",
       "      <th>Number of mutations in read 1</th>\n",
       "      <th>Number of mutations in read 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1010:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTCTTCTTAGCTCCATGATTAATTCCTAATT...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPSS</td>\n",
       "      <td>no homology found</td>\n",
       "      <td>36.509934</td>\n",
       "      <td>36.576159</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no barcode found</td>\n",
       "      <td>Y352S</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1020:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.920530</td>\n",
       "      <td>36.827815</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>GGAGGCGATTAAGAC</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1100:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.827815</td>\n",
       "      <td>36.841060</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>CCAGCCGCATCTACA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1110:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLANANVILTA...</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>36.430464</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GAGCGAGTGAGTAAG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>G383A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1180:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.841060</td>\n",
       "      <td>36.827815</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>GCAGGGCAGAATGAG</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1190:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>34.721854</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CCGTGCCGCAAAAAG</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1250:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKAVGVEDSRSALPA</td>\n",
       "      <td>32.099338</td>\n",
       "      <td>31.463576</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>TCACTTTGGGAAGGC</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>T459S</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1270:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.271523</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GTTTACTTTGCAAAC</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1320:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.827815</td>\n",
       "      <td>36.364238</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>AATGTGGGGACTACG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1360:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTCTCTGCAACCATGATTAATTCCTAATT...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFLCNHD</td>\n",
       "      <td>no homology found</td>\n",
       "      <td>36.403974</td>\n",
       "      <td>35.066225</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no barcode found</td>\n",
       "      <td>W356D</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1390:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.841060</td>\n",
       "      <td>36.589404</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CGAATAGGAGAAATA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1450:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.682119</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>ATATAGGGGAAGCAA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1490:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.827815</td>\n",
       "      <td>36.841060</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>AAGAATCCGGTATTA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1530:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.430464</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>AGGTCTATCAATGAG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1560:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.682119</td>\n",
       "      <td>36.417219</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CAAGTAGGGAATGGT</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1610:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.589404</td>\n",
       "      <td>35.503311</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>AATCCTATCGACATA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1670:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTHIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>no homology found</td>\n",
       "      <td>36.258278</td>\n",
       "      <td>34.523179</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no barcode found</td>\n",
       "      <td>P364H</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1750:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>35.582781</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CGGTAGGACAGGGAT</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1810:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.682119</td>\n",
       "      <td>34.019868</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>TGACAGCAGAGGCTA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1820:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.841060</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CACACCAGCTCACGA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1890:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>35.728477</td>\n",
       "      <td>36.589404</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>CGAGGAGACCTTAAA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1920:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.761589</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CGAGAAAGAGAGGAA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1940:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>36.112583</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>TCGGGGCGCCATCAT</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2050:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.086093</td>\n",
       "      <td>36.350993</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>CCCCTAAGACGTGAG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2100:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSAIPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.827815</td>\n",
       "      <td>36.920530</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>ACTGATGTAGAGAGC</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>L463I</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2110:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.589404</td>\n",
       "      <td>35.596026</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>AGGCATTCCACGTAA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2140:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNFLRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.761589</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CGCAGGTCCCACCGA</td>\n",
       "      <td>A372F</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2150:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGACTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td></td>\n",
       "      <td>34.576159</td>\n",
       "      <td>27.264901</td>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CCACCGCGGTAGGTC</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2160:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.841060</td>\n",
       "      <td>35.490066</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>AGGATTCTGACATCA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2190:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.682119</td>\n",
       "      <td>35.834437</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>GAGCTAATAAAGAGC</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2280:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLVNANVILTA...</td>\n",
       "      <td>33.304636</td>\n",
       "      <td>35.092715</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>TCATGGCCAGCGCGC</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Q445K</td>\n",
       "      <td>G383V</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2340:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.086093</td>\n",
       "      <td>36.920530</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GGCTGAGAAATGTAA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2350:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.509934</td>\n",
       "      <td>36.430464</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GAATGAAAGTCTAAG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2420:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFEYLGNANVILTA...</td>\n",
       "      <td>35.384106</td>\n",
       "      <td>36.589404</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>CTGGGCCCGCTCCTG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>N381Y</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2430:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAATGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDNRSALPA</td>\n",
       "      <td>33.596026</td>\n",
       "      <td>26.827815</td>\n",
       "      <td>11</td>\n",
       "      <td>42</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>ACCCCCTATCAGCTA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>T459N</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2460:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.509934</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>AAACTACCCATAGAC</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2470:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.324503</td>\n",
       "      <td>36.258278</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>AGCTTGAAATTCATT</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2490:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.761589</td>\n",
       "      <td>36.496689</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>AATAGCGGTGAGGAT</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2510:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>35.344371</td>\n",
       "      <td>36.178808</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GTGTGGCATTATACA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2570:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDIRSALPA</td>\n",
       "      <td>36.430464</td>\n",
       "      <td>35.847682</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>GCCCATTGCCCACAC</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>T459I</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2660:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.761589</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>TGCCGCCAGCCGAAC</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2710:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDSRSALPA</td>\n",
       "      <td>36.602649</td>\n",
       "      <td>35.596026</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>GTGATAGCGGCGACA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>T459S</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2780:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>TAGCGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>32.655629</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CTGGGATTCGATAAA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2800:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.827815</td>\n",
       "      <td>36.589404</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>CCCCAAACCGATCAA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2810:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>ATTGAGCCCTGGGGC</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2830:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLANANVILTA...</td>\n",
       "      <td>33.225166</td>\n",
       "      <td>34.549669</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>CCCCAAACCGATCAA</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>G383A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2900:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>35.860927</td>\n",
       "      <td>33.807947</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>AGGGCATGCGAACAG</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2930:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>35.529801</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GTGAGGAAGGTTCGT</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:2970:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.430464</td>\n",
       "      <td>35.251656</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>TGGACCTGACGTAGG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:3060:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALLVEDTRSALPA</td>\n",
       "      <td>36.364238</td>\n",
       "      <td>35.423841</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>GCCAGAAGTAAAGTG</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>G455L</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Read ID  \\\n",
       "0   FS10000719:32:BPC29518-1720:1:1101:1010:1000   \n",
       "1   FS10000719:32:BPC29518-1720:1:1101:1020:1000   \n",
       "2   FS10000719:32:BPC29518-1720:1:1101:1100:1000   \n",
       "3   FS10000719:32:BPC29518-1720:1:1101:1110:1000   \n",
       "4   FS10000719:32:BPC29518-1720:1:1101:1180:1000   \n",
       "5   FS10000719:32:BPC29518-1720:1:1101:1190:1000   \n",
       "6   FS10000719:32:BPC29518-1720:1:1101:1250:1000   \n",
       "7   FS10000719:32:BPC29518-1720:1:1101:1270:1000   \n",
       "8   FS10000719:32:BPC29518-1720:1:1101:1320:1000   \n",
       "9   FS10000719:32:BPC29518-1720:1:1101:1360:1000   \n",
       "10  FS10000719:32:BPC29518-1720:1:1101:1390:1000   \n",
       "11  FS10000719:32:BPC29518-1720:1:1101:1450:1000   \n",
       "12  FS10000719:32:BPC29518-1720:1:1101:1490:1000   \n",
       "13  FS10000719:32:BPC29518-1720:1:1101:1530:1000   \n",
       "14  FS10000719:32:BPC29518-1720:1:1101:1560:1000   \n",
       "15  FS10000719:32:BPC29518-1720:1:1101:1610:1000   \n",
       "16  FS10000719:32:BPC29518-1720:1:1101:1670:1000   \n",
       "17  FS10000719:32:BPC29518-1720:1:1101:1750:1000   \n",
       "18  FS10000719:32:BPC29518-1720:1:1101:1810:1000   \n",
       "19  FS10000719:32:BPC29518-1720:1:1101:1820:1000   \n",
       "20  FS10000719:32:BPC29518-1720:1:1101:1890:1000   \n",
       "21  FS10000719:32:BPC29518-1720:1:1101:1920:1000   \n",
       "22  FS10000719:32:BPC29518-1720:1:1101:1940:1000   \n",
       "23  FS10000719:32:BPC29518-1720:1:1101:2050:1000   \n",
       "24  FS10000719:32:BPC29518-1720:1:1101:2100:1000   \n",
       "25  FS10000719:32:BPC29518-1720:1:1101:2110:1000   \n",
       "26  FS10000719:32:BPC29518-1720:1:1101:2140:1000   \n",
       "27  FS10000719:32:BPC29518-1720:1:1101:2150:1000   \n",
       "28  FS10000719:32:BPC29518-1720:1:1101:2160:1000   \n",
       "29  FS10000719:32:BPC29518-1720:1:1101:2190:1000   \n",
       "30  FS10000719:32:BPC29518-1720:1:1101:2280:1000   \n",
       "31  FS10000719:32:BPC29518-1720:1:1101:2340:1000   \n",
       "32  FS10000719:32:BPC29518-1720:1:1101:2350:1000   \n",
       "33  FS10000719:32:BPC29518-1720:1:1101:2420:1000   \n",
       "34  FS10000719:32:BPC29518-1720:1:1101:2430:1000   \n",
       "35  FS10000719:32:BPC29518-1720:1:1101:2460:1000   \n",
       "36  FS10000719:32:BPC29518-1720:1:1101:2470:1000   \n",
       "37  FS10000719:32:BPC29518-1720:1:1101:2490:1000   \n",
       "38  FS10000719:32:BPC29518-1720:1:1101:2510:1000   \n",
       "39  FS10000719:32:BPC29518-1720:1:1101:2570:1000   \n",
       "40  FS10000719:32:BPC29518-1720:1:1101:2660:1000   \n",
       "41  FS10000719:32:BPC29518-1720:1:1101:2710:1000   \n",
       "42  FS10000719:32:BPC29518-1720:1:1101:2780:1000   \n",
       "43  FS10000719:32:BPC29518-1720:1:1101:2800:1000   \n",
       "44  FS10000719:32:BPC29518-1720:1:1101:2810:1000   \n",
       "45  FS10000719:32:BPC29518-1720:1:1101:2830:1000   \n",
       "46  FS10000719:32:BPC29518-1720:1:1101:2900:1000   \n",
       "47  FS10000719:32:BPC29518-1720:1:1101:2930:1000   \n",
       "48  FS10000719:32:BPC29518-1720:1:1101:2970:1000   \n",
       "49  FS10000719:32:BPC29518-1720:1:1101:3060:1000   \n",
       "\n",
       "                                            Sequence1  \\\n",
       "0   TGAAGCACAAGGCCCCTCTTCTTAGCTCCATGATTAATTCCTAATT...   \n",
       "1   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "2   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "3   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "4   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "5   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "6   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "7   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "8   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "9   TGAAGCACAAGGCCCCTTTCTCTGCAACCATGATTAATTCCTAATT...   \n",
       "10  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "11  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "12  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "13  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "14  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "15  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "16  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "17  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "18  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "19  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "20  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "21  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "22  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "23  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "24  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "25  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "26  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "27  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "28  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "29  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "30  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "31  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "32  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "33  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "34  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "35  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "36  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "37  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "38  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "39  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "40  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "41  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "42  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "43  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "44  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "45  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "46  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "47  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "48  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "49  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "\n",
       "                                            Sequence2  \\\n",
       "0   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "1   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "2   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "3   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "4   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "5   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "6   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "7   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "8   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "9   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "10  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "11  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "12  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "13  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "14  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "15  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "16  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "17  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "18  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "19  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "20  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "21  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "22  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "23  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "24  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "25  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "26  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "27  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGACTGTCAACAAA...   \n",
       "28  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "29  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "30  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "31  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "32  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "33  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "34  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAATGTCAACAAA...   \n",
       "35  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "36  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "37  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "38  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "39  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "40  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "41  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "42  TAGCGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAAA...   \n",
       "43  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "44  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "45  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "46  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "47  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "48  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "49  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "\n",
       "                                         translation1  \\\n",
       "0                                             EAQGPSS   \n",
       "1   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "2                                    WRKALGVEDTRSALPA   \n",
       "3                                    WRKALGVEDTRSALPA   \n",
       "4   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "5   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "6   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "7                                    WRKALGVEDTRSALPA   \n",
       "8                                    WRKALGVEDTRSALPA   \n",
       "9                                         EAQGPFLCNHD   \n",
       "10  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "11                                   WRKALGVEDTRSALPA   \n",
       "12                                   WRKALGVEDTRSALPA   \n",
       "13                                   WRKALGVEDTRSALPA   \n",
       "14  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "15                                   WRKALGVEDTRSALPA   \n",
       "16  EAQGPFYRQSWGGMKACTHIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "17  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "18  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "19  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "20                                   WRKALGVEDTRSALPA   \n",
       "21  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "22                                   WRKALGVEDTRSALPA   \n",
       "23                                   WRKALGVEDTRSALPA   \n",
       "24                                   WRKALGVEDTRSAIPA   \n",
       "25  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "26  EAQGPFYRQSWGGMKACTPIISGGMNFLRMPGFFENLGNANVILTA...   \n",
       "27  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "28  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "29  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "30                                   WRKALGVEDTRSALPA   \n",
       "31                                   WRKALGVEDTRSALPA   \n",
       "32                                   WRKALGVEDTRSALPA   \n",
       "33                                   WRKALGVEDTRSALPA   \n",
       "34  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "35                                   WRKALGVEDTRSALPA   \n",
       "36                                   WRKALGVEDTRSALPA   \n",
       "37  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "38                                   WRKALGVEDTRSALPA   \n",
       "39  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "40                                   WRKALGVEDTRSALPA   \n",
       "41  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "42  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "43                                   WRKALGVEDTRSALPA   \n",
       "44                                   WRKALGVEDTRSALPA   \n",
       "45                                   WRKALGVEDTRSALPA   \n",
       "46  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "47                                   WRKALGVEDTRSALPA   \n",
       "48                                   WRKALGVEDTRSALPA   \n",
       "49  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "\n",
       "                                         translation2  Average phredQ1  \\\n",
       "0                                   no homology found        36.509934   \n",
       "1                                    WRKALGVEDTRSALPA        36.920530   \n",
       "2   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.827815   \n",
       "3   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLANANVILTA...        36.668874   \n",
       "4                                    WRKALGVEDTRSALPA        36.841060   \n",
       "5                                    WRKALGVEDTRSALPA        36.748344   \n",
       "6                                    WRKAVGVEDSRSALPA        32.099338   \n",
       "7   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "8   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.827815   \n",
       "9                                   no homology found        36.403974   \n",
       "10                                   WRKALGVEDTRSALPA        36.841060   \n",
       "11  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "12  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.827815   \n",
       "13  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "14                                   WRKALGVEDTRSALPA        36.682119   \n",
       "15  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.589404   \n",
       "16                                  no homology found        36.258278   \n",
       "17                                   WRKALGVEDTRSALPA        36.668874   \n",
       "18                                   WRKALGVEDTRSALPA        36.682119   \n",
       "19                                   WRKALGVEDTRSALPA        36.841060   \n",
       "20  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        35.728477   \n",
       "21                                   WRKALGVEDTRSALPA        36.761589   \n",
       "22  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.668874   \n",
       "23  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.086093   \n",
       "24  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.827815   \n",
       "25                                   WRKALGVEDTRSALPA        36.589404   \n",
       "26                                   WRKALGVEDTRSALPA        36.761589   \n",
       "27                                                           34.576159   \n",
       "28                                   WRKALGVEDTRSALPA        36.841060   \n",
       "29                                   WRKALGVEDTRSALPA        36.682119   \n",
       "30  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLVNANVILTA...        33.304636   \n",
       "31  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.086093   \n",
       "32  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.509934   \n",
       "33  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFEYLGNANVILTA...        35.384106   \n",
       "34                                   WRKALGVEDNRSALPA        33.596026   \n",
       "35  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "36  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.324503   \n",
       "37                                   WRKALGVEDTRSALPA        36.761589   \n",
       "38  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        35.344371   \n",
       "39                                   WRKALGVEDIRSALPA        36.430464   \n",
       "40  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "41                                   WRKALGVEDSRSALPA        36.602649   \n",
       "42                                   WRKALGVEDTRSALPA        36.668874   \n",
       "43  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.827815   \n",
       "44  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "45  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLANANVILTA...        33.225166   \n",
       "46                                   WRKALGVEDTRSALPA        35.860927   \n",
       "47  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.668874   \n",
       "48  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.430464   \n",
       "49                                   WRKALLVEDTRSALPA        36.364238   \n",
       "\n",
       "    Average phredQ2  Number of bases with phredQ under 20 read 1  \\\n",
       "0         36.576159                                            1   \n",
       "1         36.827815                                            0   \n",
       "2         36.841060                                            1   \n",
       "3         36.430464                                            1   \n",
       "4         36.827815                                            0   \n",
       "5         34.721854                                            1   \n",
       "6         31.463576                                           16   \n",
       "7         36.271523                                            1   \n",
       "8         36.364238                                            1   \n",
       "9         35.066225                                            3   \n",
       "10        36.589404                                            0   \n",
       "11        36.682119                                            1   \n",
       "12        36.841060                                            1   \n",
       "13        36.430464                                            1   \n",
       "14        36.417219                                            0   \n",
       "15        35.503311                                            1   \n",
       "16        34.523179                                            2   \n",
       "17        35.582781                                            1   \n",
       "18        34.019868                                            0   \n",
       "19        36.668874                                            0   \n",
       "20        36.589404                                            6   \n",
       "21        36.748344                                            0   \n",
       "22        36.112583                                            1   \n",
       "23        36.350993                                            3   \n",
       "24        36.920530                                            1   \n",
       "25        35.596026                                            1   \n",
       "26        36.668874                                            0   \n",
       "27        27.264901                                            9   \n",
       "28        35.490066                                            0   \n",
       "29        35.834437                                            0   \n",
       "30        35.092715                                           15   \n",
       "31        36.920530                                            3   \n",
       "32        36.430464                                            1   \n",
       "33        36.589404                                            8   \n",
       "34        26.827815                                           11   \n",
       "35        36.509934                                            1   \n",
       "36        36.258278                                            3   \n",
       "37        36.496689                                            0   \n",
       "38        36.178808                                            5   \n",
       "39        35.847682                                            1   \n",
       "40        36.761589                                            1   \n",
       "41        35.596026                                            0   \n",
       "42        32.655629                                            1   \n",
       "43        36.589404                                            1   \n",
       "44        36.668874                                            1   \n",
       "45        34.549669                                            9   \n",
       "46        33.807947                                            2   \n",
       "47        35.529801                                            1   \n",
       "48        35.251656                                            1   \n",
       "49        35.423841                                            0   \n",
       "\n",
       "    Number of bases with phredQ under 20 read 2  \\\n",
       "0                                             2   \n",
       "1                                             1   \n",
       "2                                             0   \n",
       "3                                             1   \n",
       "4                                             1   \n",
       "5                                            10   \n",
       "6                                            22   \n",
       "7                                             1   \n",
       "8                                             0   \n",
       "9                                             2   \n",
       "10                                            1   \n",
       "11                                            0   \n",
       "12                                            0   \n",
       "13                                            1   \n",
       "14                                            2   \n",
       "15                                            5   \n",
       "16                                            7   \n",
       "17                                            5   \n",
       "18                                            9   \n",
       "19                                            1   \n",
       "20                                            1   \n",
       "21                                            1   \n",
       "22                                            1   \n",
       "23                                            1   \n",
       "24                                            0   \n",
       "25                                            4   \n",
       "26                                            1   \n",
       "27                                           45   \n",
       "28                                            6   \n",
       "29                                            4   \n",
       "30                                            6   \n",
       "31                                            0   \n",
       "32                                            1   \n",
       "33                                            1   \n",
       "34                                           42   \n",
       "35                                            1   \n",
       "36                                            2   \n",
       "37                                            2   \n",
       "38                                            2   \n",
       "39                                            3   \n",
       "40                                            0   \n",
       "41                                            4   \n",
       "42                                           16   \n",
       "43                                            1   \n",
       "44                                            1   \n",
       "45                                            5   \n",
       "46                                            7   \n",
       "47                                            3   \n",
       "48                                            6   \n",
       "49                                            5   \n",
       "\n",
       "    Is there a barcode in read 1?  Is there a barcode in read 2?  \\\n",
       "0                           False                          False   \n",
       "1                           False                           True   \n",
       "2                            True                          False   \n",
       "3                            True                          False   \n",
       "4                           False                           True   \n",
       "5                           False                           True   \n",
       "6                           False                           True   \n",
       "7                            True                          False   \n",
       "8                            True                          False   \n",
       "9                           False                          False   \n",
       "10                          False                           True   \n",
       "11                           True                          False   \n",
       "12                           True                          False   \n",
       "13                           True                          False   \n",
       "14                          False                           True   \n",
       "15                           True                          False   \n",
       "16                          False                          False   \n",
       "17                          False                           True   \n",
       "18                          False                           True   \n",
       "19                          False                           True   \n",
       "20                           True                          False   \n",
       "21                          False                           True   \n",
       "22                           True                          False   \n",
       "23                           True                          False   \n",
       "24                           True                          False   \n",
       "25                          False                           True   \n",
       "26                          False                           True   \n",
       "27                          False                           True   \n",
       "28                          False                           True   \n",
       "29                          False                           True   \n",
       "30                           True                          False   \n",
       "31                           True                          False   \n",
       "32                           True                          False   \n",
       "33                           True                          False   \n",
       "34                          False                           True   \n",
       "35                           True                          False   \n",
       "36                           True                          False   \n",
       "37                          False                           True   \n",
       "38                           True                          False   \n",
       "39                          False                           True   \n",
       "40                           True                          False   \n",
       "41                          False                           True   \n",
       "42                          False                           True   \n",
       "43                           True                          False   \n",
       "44                           True                          False   \n",
       "45                           True                          False   \n",
       "46                          False                           True   \n",
       "47                           True                          False   \n",
       "48                           True                          False   \n",
       "49                          False                           True   \n",
       "\n",
       "    Nonbarcode read start position1  Nonbarcode read start position2  \\\n",
       "0                              1037                               -1   \n",
       "1                              1037                               -1   \n",
       "2                                -1                             1037   \n",
       "3                                -1                             1037   \n",
       "4                              1037                               -1   \n",
       "5                              1037                               -1   \n",
       "6                              1037                               -1   \n",
       "7                                -1                             1037   \n",
       "8                                -1                             1037   \n",
       "9                              1037                               -1   \n",
       "10                             1037                               -1   \n",
       "11                               -1                             1037   \n",
       "12                               -1                             1037   \n",
       "13                               -1                             1037   \n",
       "14                             1037                               -1   \n",
       "15                               -1                             1037   \n",
       "16                             1037                               -1   \n",
       "17                             1037                               -1   \n",
       "18                             1037                               -1   \n",
       "19                             1037                               -1   \n",
       "20                               -1                             1037   \n",
       "21                             1037                               -1   \n",
       "22                               -1                             1037   \n",
       "23                               -1                             1037   \n",
       "24                               -1                             1037   \n",
       "25                             1037                               -1   \n",
       "26                             1037                               -1   \n",
       "27                             1037                               -1   \n",
       "28                             1037                               -1   \n",
       "29                             1037                               -1   \n",
       "30                               -1                             1037   \n",
       "31                               -1                             1037   \n",
       "32                               -1                             1037   \n",
       "33                               -1                             1037   \n",
       "34                             1037                               -1   \n",
       "35                               -1                             1037   \n",
       "36                               -1                             1037   \n",
       "37                             1037                               -1   \n",
       "38                               -1                             1037   \n",
       "39                             1037                               -1   \n",
       "40                               -1                             1037   \n",
       "41                             1037                               -1   \n",
       "42                             1037                               -1   \n",
       "43                               -1                             1037   \n",
       "44                               -1                             1037   \n",
       "45                               -1                             1037   \n",
       "46                             1037                               -1   \n",
       "47                               -1                             1037   \n",
       "48                               -1                             1037   \n",
       "49                             1037                               -1   \n",
       "\n",
       "        Barcode sequence1      Barcode sequence2    Mutation1    Mutation2  \\\n",
       "0   valid nonbarcode read       no barcode found        Y352S  no mutation   \n",
       "1   valid nonbarcode read        GGAGGCGATTAAGAC  no mutation  no mutation   \n",
       "2         CCAGCCGCATCTACA  valid nonbarcode read  no mutation  no mutation   \n",
       "3         GAGCGAGTGAGTAAG  valid nonbarcode read  no mutation        G383A   \n",
       "4   valid nonbarcode read        GCAGGGCAGAATGAG  no mutation  no mutation   \n",
       "5   valid nonbarcode read        CCGTGCCGCAAAAAG  no mutation        Y447N   \n",
       "6   valid nonbarcode read        TCACTTTGGGAAGGC  no mutation        T459S   \n",
       "7         GTTTACTTTGCAAAC  valid nonbarcode read        Y447N  no mutation   \n",
       "8         AATGTGGGGACTACG  valid nonbarcode read  no mutation  no mutation   \n",
       "9   valid nonbarcode read       no barcode found        W356D  no mutation   \n",
       "10  valid nonbarcode read        CGAATAGGAGAAATA  no mutation        Y447N   \n",
       "11        ATATAGGGGAAGCAA  valid nonbarcode read  no mutation  no mutation   \n",
       "12        AAGAATCCGGTATTA  valid nonbarcode read  no mutation  no mutation   \n",
       "13        AGGTCTATCAATGAG  valid nonbarcode read        Y447N  no mutation   \n",
       "14  valid nonbarcode read        CAAGTAGGGAATGGT  no mutation        Y447N   \n",
       "15        AATCCTATCGACATA  valid nonbarcode read        Y447N  no mutation   \n",
       "16  valid nonbarcode read       no barcode found        P364H  no mutation   \n",
       "17  valid nonbarcode read        CGGTAGGACAGGGAT  no mutation        Y447N   \n",
       "18  valid nonbarcode read        TGACAGCAGAGGCTA  no mutation  no mutation   \n",
       "19  valid nonbarcode read        CACACCAGCTCACGA  no mutation  no mutation   \n",
       "20        CGAGGAGACCTTAAA  valid nonbarcode read        Y447N  no mutation   \n",
       "21  valid nonbarcode read        CGAGAAAGAGAGGAA  no mutation  no mutation   \n",
       "22        TCGGGGCGCCATCAT  valid nonbarcode read  no mutation  no mutation   \n",
       "23        CCCCTAAGACGTGAG  valid nonbarcode read        Y447N  no mutation   \n",
       "24        ACTGATGTAGAGAGC  valid nonbarcode read        L463I  no mutation   \n",
       "25  valid nonbarcode read        AGGCATTCCACGTAA  no mutation        Y447N   \n",
       "26  valid nonbarcode read        CGCAGGTCCCACCGA        A372F  no mutation   \n",
       "27  valid nonbarcode read        CCACCGCGGTAGGTC  no mutation        Y447N   \n",
       "28  valid nonbarcode read        AGGATTCTGACATCA  no mutation  no mutation   \n",
       "29  valid nonbarcode read        GAGCTAATAAAGAGC  no mutation        Y447N   \n",
       "30        TCATGGCCAGCGCGC  valid nonbarcode read        Q445K        G383V   \n",
       "31        GGCTGAGAAATGTAA  valid nonbarcode read  no mutation  no mutation   \n",
       "32        GAATGAAAGTCTAAG  valid nonbarcode read        Y447N  no mutation   \n",
       "33        CTGGGCCCGCTCCTG  valid nonbarcode read  no mutation        N381Y   \n",
       "34  valid nonbarcode read        ACCCCCTATCAGCTA  no mutation        T459N   \n",
       "35        AAACTACCCATAGAC  valid nonbarcode read  no mutation  no mutation   \n",
       "36        AGCTTGAAATTCATT  valid nonbarcode read        Y447N  no mutation   \n",
       "37  valid nonbarcode read        AATAGCGGTGAGGAT  no mutation  no mutation   \n",
       "38        GTGTGGCATTATACA  valid nonbarcode read        Y447N  no mutation   \n",
       "39  valid nonbarcode read        GCCCATTGCCCACAC  no mutation        T459I   \n",
       "40        TGCCGCCAGCCGAAC  valid nonbarcode read  no mutation  no mutation   \n",
       "41  valid nonbarcode read        GTGATAGCGGCGACA  no mutation        T459S   \n",
       "42  valid nonbarcode read        CTGGGATTCGATAAA  no mutation  no mutation   \n",
       "43        CCCCAAACCGATCAA  valid nonbarcode read  no mutation  no mutation   \n",
       "44        ATTGAGCCCTGGGGC  valid nonbarcode read        Y447N  no mutation   \n",
       "45        CCCCAAACCGATCAA  valid nonbarcode read        Y447N        G383A   \n",
       "46  valid nonbarcode read        AGGGCATGCGAACAG  no mutation  no mutation   \n",
       "47        GTGAGGAAGGTTCGT  valid nonbarcode read        Y447N  no mutation   \n",
       "48        TGGACCTGACGTAGG  valid nonbarcode read  no mutation  no mutation   \n",
       "49  valid nonbarcode read        GCCAGAAGTAAAGTG  no mutation        G455L   \n",
       "\n",
       "    Number of mutations in read 1  Number of mutations in read 2  \n",
       "0                               2                              0  \n",
       "1                               0                              0  \n",
       "2                               0                              0  \n",
       "3                               0                              1  \n",
       "4                               0                              0  \n",
       "5                               0                              1  \n",
       "6                               0                              3  \n",
       "7                               1                              0  \n",
       "8                               0                              0  \n",
       "9                               5                              0  \n",
       "10                              0                              1  \n",
       "11                              0                              0  \n",
       "12                              0                              0  \n",
       "13                              1                              0  \n",
       "14                              0                              1  \n",
       "15                              1                              0  \n",
       "16                              1                              0  \n",
       "17                              0                              1  \n",
       "18                              0                              0  \n",
       "19                              0                              0  \n",
       "20                              1                              0  \n",
       "21                              0                              0  \n",
       "22                              0                              0  \n",
       "23                              1                              0  \n",
       "24                              1                              0  \n",
       "25                              0                              1  \n",
       "26                              1                              0  \n",
       "27                              0                              1  \n",
       "28                              0                              0  \n",
       "29                              0                              1  \n",
       "30                              1                              1  \n",
       "31                              0                              0  \n",
       "32                              1                              0  \n",
       "33                              0                              1  \n",
       "34                              0                              2  \n",
       "35                              0                              0  \n",
       "36                              1                              0  \n",
       "37                              0                              0  \n",
       "38                              1                              0  \n",
       "39                              0                              1  \n",
       "40                              0                              0  \n",
       "41                              0                              2  \n",
       "42                              0                              0  \n",
       "43                              0                              0  \n",
       "44                              1                              0  \n",
       "45                              1                              1  \n",
       "46                              0                              0  \n",
       "47                              1                              0  \n",
       "48                              0                              0  \n",
       "49                              0                              2  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# here I convert a SeqIO file into a pandas dataframe so that I can do statistics\n",
    "# First I parse the fastq file, then I make a dictionary with the info I want then I make a dataframe\n",
    "# I want to make a new column that identifies if there is a mutation and what it is\n",
    "# Also I'll need to pair paired reads somehow (http://bytesizebio.net/2011/11/10/short-bioinformatics-hacks-reading-mate-pairs-from-a-fastq-file/)\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "import math\n",
    "\n",
    "def processRecord(rec, sequences, AVGphredQ, basesWphredQundertwenty, translations, barcodePresence,\n",
    "                  barcodeList, nonBarcodeReadAlignment, mutation, numberOfMutations, readIDs = None):\n",
    "    if readIDs is not None:\n",
    "        readIDs.append(rec.id)\n",
    "    sequences.append(str(rec.seq))\n",
    "    AVGphredQ.append(np.mean(rec.letter_annotations['phred_quality']))\n",
    "    badReads = 0\n",
    "    for qualityScore in rec.letter_annotations['phred_quality']:\n",
    "        if qualityScore < 20:\n",
    "            badReads += 1\n",
    "    basesWphredQundertwenty.append(badReads)\n",
    "    barcodeStart = rec.seq.find(BARCODE_FLANK)\n",
    "    mutant = 'no mutation'\n",
    "    \n",
    "    # This finds barcodes and translations\n",
    "    # I'm leaving a bit of this translation on the table but this explicitly gets it in frame\n",
    "\n",
    "    if barcodeStart != -1:\n",
    "        # first get a translation of the region next to the barcode\n",
    "        translations.append(str(rec.seq[barcodeStart + 14:barcodeStart + 62].reverse_complement().translate(to_stop=True)))\n",
    "         \n",
    "        # if there is a barcode add a column saying that there was one\n",
    "        barcodeToAppend = str(rec.seq[barcodeStart - 15:barcodeStart])\n",
    "        if barcodeToAppend == '' or barcodeToAppend == None or len(barcodeToAppend) != 15:\n",
    "            barcodeList.append('no barcode found')\n",
    "            barcodePresence.append(False)\n",
    "        else:\n",
    "            barcodeList.append(barcodeToAppend)\n",
    "            barcodePresence.append(True)\n",
    "        #this collects the barcode sequence\n",
    "        nonBarcodeReadAlignment.append(-1) \n",
    "        # This just fills in an empty space for the alignment of the nonbarcode read\n",
    "        mutationCounter = 0\n",
    "        # The start of the enumeration should be the end of the protein sequence, in this case \n",
    "        # it is as long as the barcode flank\n",
    "        stopCodonPosition = barcodeStart + len(BARCODE_FLANK)\n",
    "        lenOfCodingSequence = len(rec.seq) - stopCodonPosition\n",
    "        lengthInFrame = lenOfCodingSequence - (lenOfCodingSequence % 3)\n",
    "        barcodeReadStart = len(TEMPLATE_TRANSLATION) - (lengthInFrame // 3)\n",
    "        for position, aminoAcid in enumerate(str((rec.seq[stopCodonPosition:stopCodonPosition + lengthInFrame].reverse_complement().translate(to_stop = True)))):\n",
    "            templateAminoAcid = TEMPLATE_TRANSLATION[barcodeReadStart + position]\n",
    "            if aminoAcid != templateAminoAcid:\n",
    "                mutant = str(templateAminoAcid) + str(barcodeReadStart + position) + str(aminoAcid)\n",
    "                mutationCounter += 1\n",
    "        mutation.append(mutant)\n",
    "        numberOfMutations.append(mutationCounter)\n",
    "        return\n",
    "    \n",
    "    nonBarcodeReadStart = TEMPLATE_STRING.find(str(rec.seq[:SEARCH_MOTIF_LENGTH])) \n",
    "\n",
    "    # this finds the read in the template seq\n",
    "    if nonBarcodeReadStart != -1:\n",
    "        barcodePresence.append(False) \n",
    "        # this indicates that no barcode was found\n",
    "        barcodeList.append('valid nonbarcode read')\n",
    "        # this finds the frame\n",
    "        frame = nonBarcodeReadStart % 3  \n",
    "        # I have to subtract the frame to figure out where to start translating\n",
    "        nonBarcodeReadingFrame = rec.seq[3 - frame:] \n",
    "        translations.append(str(nonBarcodeReadingFrame.translate(to_stop = True))) \n",
    "        # This adds in where the alignment begins in the DNA\n",
    "        nonBarcodeReadAlignment.append(nonBarcodeReadStart) \n",
    "        # This adds in where the alignment begins in the protein\n",
    "        nonBarcodeTranslationStart = math.floor(nonBarcodeReadStart / 3) + 1 \n",
    "        # Here I will compare the template translation to the read translation\n",
    "        mutationCounter = 0\n",
    "        for position, aminoAcid in enumerate(str(nonBarcodeReadingFrame.translate(to_stop = True))):\n",
    "            templateIndex = nonBarcodeTranslationStart + position\n",
    "            if templateIndex >= len(TEMPLATE_TRANSLATION):\n",
    "                break\n",
    "#             print('the index is ' + str(templateIndex))\n",
    "            templateAminoAcid = TEMPLATE_TRANSLATION[templateIndex]\n",
    "            if aminoAcid != templateAminoAcid:\n",
    "                mutant = str(templateAminoAcid) + str(templateIndex) + str(aminoAcid)\n",
    "                mutationCounter += 1\n",
    "        mutation.append(mutant)\n",
    "        numberOfMutations.append(mutationCounter)\n",
    "        return\n",
    "    \n",
    "    # if no barcode or other homolgy is found this takes care of failures\n",
    "    translations.append('no homology found')\n",
    "    barcodePresence.append(False)\n",
    "    barcodeList.append('no barcode found')\n",
    "    nonBarcodeReadAlignment.append(-1)\n",
    "    mutation.append(mutant)\n",
    "    numberOfMutations.append(0)\n",
    "\n",
    "# Here I can import any fastq file with reads\n",
    "mockDataParsed = SeqIO.parse(\"NP-11-19-4-merged.fastq\", \"fastq\")\n",
    "\n",
    "readIDs = []\n",
    "sequences1 = []\n",
    "sequences2 = []\n",
    "translations1 = []\n",
    "translations2 = []\n",
    "AVGphredQ1 = []\n",
    "AVGphredQ2 = []\n",
    "basesWphredQundertwenty1 = []\n",
    "basesWphredQundertwenty2 = []\n",
    "barcodePresence1 = []\n",
    "barcodePresence2 = []\n",
    "nonBarcodeReadAlignment1 = []\n",
    "nonBarcodeReadAlignment2 = []\n",
    "barcodeList1 = []\n",
    "barcodeList2 = []\n",
    "mutation1 = []\n",
    "mutation2 = []\n",
    "numberOfMutations1 = []\n",
    "numberOfMutations2 = []\n",
    "\n",
    "#I'm going to put paired reads into the same row in this dictionary\n",
    "#The file will have to have an even number of reads for this\n",
    "\n",
    "for rec1, rec2 in itertools.zip_longest(mockDataParsed, mockDataParsed): \n",
    "    processRecord(rec1, sequences1, AVGphredQ1, basesWphredQundertwenty1, translations1, \n",
    "                  barcodePresence1, barcodeList1, nonBarcodeReadAlignment1, mutation1, numberOfMutations1, readIDs)\n",
    "    processRecord(rec2, sequences2, AVGphredQ2, basesWphredQundertwenty2, translations2, \n",
    "                  barcodePresence2, barcodeList2, nonBarcodeReadAlignment2, mutation2, numberOfMutations2)\n",
    "\n",
    "untrimmed_data = {\n",
    "    'Read ID': readIDs,\n",
    "    'Sequence1': sequences1,\n",
    "    'Sequence2': sequences2,\n",
    "    'translation1': translations1,\n",
    "    'translation2': translations2,\n",
    "    'Average phredQ1': AVGphredQ1,\n",
    "    'Average phredQ2': AVGphredQ2,\n",
    "    'Number of bases with phredQ under 20 read 1': basesWphredQundertwenty1,\n",
    "    'Number of bases with phredQ under 20 read 2': basesWphredQundertwenty2,\n",
    "    'Is there a barcode in read 1?' : barcodePresence1,\n",
    "    'Is there a barcode in read 2?' : barcodePresence2,\n",
    "    'Nonbarcode read start position1' : nonBarcodeReadAlignment1,\n",
    "    'Nonbarcode read start position2' : nonBarcodeReadAlignment2,\n",
    "    'Barcode sequence1' : barcodeList1,\n",
    "    'Barcode sequence2' : barcodeList2,\n",
    "    'Mutation1' : mutation1,\n",
    "    'Mutation2' : mutation2,\n",
    "    'Number of mutations in read 1' : numberOfMutations1,\n",
    "    'Number of mutations in read 2' : numberOfMutations2,\n",
    "                 }\n",
    "\n",
    "dfRaw = pd.DataFrame(untrimmed_data)\n",
    "\n",
    "dfRaw.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Single mutant counts: \n",
      "Single mutant barcode counts: \n",
      "Total mutant barcode counts: \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Read ID</th>\n",
       "      <th>Sequence1</th>\n",
       "      <th>Sequence2</th>\n",
       "      <th>translation1</th>\n",
       "      <th>translation2</th>\n",
       "      <th>Average phredQ1</th>\n",
       "      <th>Average phredQ2</th>\n",
       "      <th>Number of bases with phredQ under 20 read 1</th>\n",
       "      <th>Number of bases with phredQ under 20 read 2</th>\n",
       "      <th>Is there a barcode in read 1?</th>\n",
       "      <th>...</th>\n",
       "      <th>Nonbarcode read start position1</th>\n",
       "      <th>Nonbarcode read start position2</th>\n",
       "      <th>Barcode sequence1</th>\n",
       "      <th>Barcode sequence2</th>\n",
       "      <th>Mutation1</th>\n",
       "      <th>Mutation2</th>\n",
       "      <th>Number of mutations in read 1</th>\n",
       "      <th>Number of mutations in read 2</th>\n",
       "      <th>Barcode</th>\n",
       "      <th>Single mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1110:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLANANVILTA...</td>\n",
       "      <td>36.668874</td>\n",
       "      <td>36.430464</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GAGCGAGTGAGTAAG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>G383A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGCGAGTGAGTAAG</td>\n",
       "      <td>G383A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1270:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.271523</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>GTTTACTTTGCAAAC</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>GTTTACTTTGCAAAC</td>\n",
       "      <td>Y447N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1390:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.841060</td>\n",
       "      <td>36.589404</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CGAATAGGAGAAATA</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>CGAATAGGAGAAATA</td>\n",
       "      <td>Y447N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1530:1000</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>36.748344</td>\n",
       "      <td>36.430464</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>1037</td>\n",
       "      <td>AGGTCTATCAATGAG</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>AGGTCTATCAATGAG</td>\n",
       "      <td>Y447N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FS10000719:32:BPC29518-1720:1:1101:1560:1000</td>\n",
       "      <td>TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...</td>\n",
       "      <td>ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...</td>\n",
       "      <td>EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...</td>\n",
       "      <td>WRKALGVEDTRSALPA</td>\n",
       "      <td>36.682119</td>\n",
       "      <td>36.417219</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>1037</td>\n",
       "      <td>-1</td>\n",
       "      <td>valid nonbarcode read</td>\n",
       "      <td>CAAGTAGGGAATGGT</td>\n",
       "      <td>no mutation</td>\n",
       "      <td>Y447N</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>CAAGTAGGGAATGGT</td>\n",
       "      <td>Y447N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Read ID  \\\n",
       "3   FS10000719:32:BPC29518-1720:1:1101:1110:1000   \n",
       "7   FS10000719:32:BPC29518-1720:1:1101:1270:1000   \n",
       "10  FS10000719:32:BPC29518-1720:1:1101:1390:1000   \n",
       "13  FS10000719:32:BPC29518-1720:1:1101:1530:1000   \n",
       "14  FS10000719:32:BPC29518-1720:1:1101:1560:1000   \n",
       "\n",
       "                                            Sequence1  \\\n",
       "3   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "7   ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "10  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "13  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "14  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "\n",
       "                                            Sequence2  \\\n",
       "3   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "7   TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "10  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "13  TGAAGCACAAGGCCCCTTTTATCGCCAGTCGTGGGGAGGGATGAAG...   \n",
       "14  ATAGGGAGTGGTAAAATAACTCTATCAATGATAGAGTGTCAACAAA...   \n",
       "\n",
       "                                         translation1  \\\n",
       "3                                    WRKALGVEDTRSALPA   \n",
       "7                                    WRKALGVEDTRSALPA   \n",
       "10  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "13                                   WRKALGVEDTRSALPA   \n",
       "14  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...   \n",
       "\n",
       "                                         translation2  Average phredQ1  \\\n",
       "3   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLANANVILTA...        36.668874   \n",
       "7   EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "10                                   WRKALGVEDTRSALPA        36.841060   \n",
       "13  EAQGPFYRQSWGGMKACTPIISGGMNALRMPGFFENLGNANVILTA...        36.748344   \n",
       "14                                   WRKALGVEDTRSALPA        36.682119   \n",
       "\n",
       "    Average phredQ2  Number of bases with phredQ under 20 read 1  \\\n",
       "3         36.430464                                            1   \n",
       "7         36.271523                                            1   \n",
       "10        36.589404                                            0   \n",
       "13        36.430464                                            1   \n",
       "14        36.417219                                            0   \n",
       "\n",
       "    Number of bases with phredQ under 20 read 2  \\\n",
       "3                                             1   \n",
       "7                                             1   \n",
       "10                                            1   \n",
       "13                                            1   \n",
       "14                                            2   \n",
       "\n",
       "    Is there a barcode in read 1?  ...  Nonbarcode read start position1  \\\n",
       "3                            True  ...                               -1   \n",
       "7                            True  ...                               -1   \n",
       "10                          False  ...                             1037   \n",
       "13                           True  ...                               -1   \n",
       "14                          False  ...                             1037   \n",
       "\n",
       "    Nonbarcode read start position2      Barcode sequence1  \\\n",
       "3                              1037        GAGCGAGTGAGTAAG   \n",
       "7                              1037        GTTTACTTTGCAAAC   \n",
       "10                               -1  valid nonbarcode read   \n",
       "13                             1037        AGGTCTATCAATGAG   \n",
       "14                               -1  valid nonbarcode read   \n",
       "\n",
       "        Barcode sequence2    Mutation1    Mutation2  \\\n",
       "3   valid nonbarcode read  no mutation        G383A   \n",
       "7   valid nonbarcode read        Y447N  no mutation   \n",
       "10        CGAATAGGAGAAATA  no mutation        Y447N   \n",
       "13  valid nonbarcode read        Y447N  no mutation   \n",
       "14        CAAGTAGGGAATGGT  no mutation        Y447N   \n",
       "\n",
       "   Number of mutations in read 1  Number of mutations in read 2  \\\n",
       "3                              0                              1   \n",
       "7                              1                              0   \n",
       "10                             0                              1   \n",
       "13                             1                              0   \n",
       "14                             0                              1   \n",
       "\n",
       "            Barcode Single mutation  \n",
       "3   GAGCGAGTGAGTAAG           G383A  \n",
       "7   GTTTACTTTGCAAAC           Y447N  \n",
       "10  CGAATAGGAGAAATA           Y447N  \n",
       "13  AGGTCTATCAATGAG           Y447N  \n",
       "14  CAAGTAGGGAATGGT           Y447N  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pull out rows (paired reads) that have one barcode in one of the reads and only 1 mutation between the reads\n",
    "\n",
    "i = 0\n",
    "\n",
    "mutantRows = []\n",
    "WTRows = []\n",
    "WTorSingleRow = []\n",
    "qualityTrimmed = []\n",
    "\n",
    "# Make a new column in the dataframe that has a barcode from the two barcode columns\n",
    "dummyBarcode = 'Barcode was not added'\n",
    "dfRaw['Barcode'] = dummyBarcode\n",
    "dfRaw['Barcode'] = np.where(\n",
    "    (dfRaw['Barcode sequence1'] == 'no barcode found') | (dfRaw['Barcode sequence1'] == 'valid nonbarcode read'), \n",
    "    dfRaw['Barcode sequence2'], \n",
    "    dfRaw['Barcode sequence1'])\n",
    "\n",
    "# Trim away low quality reads\n",
    "averageQualityCutoff = 25\n",
    "maxBadReads = 5\n",
    "for index, row in dfRaw.iterrows():\n",
    "    if row['Average phredQ1'] > averageQualityCutoff and row['Average phredQ2'] > averageQualityCutoff and \\\n",
    "        row['Number of bases with phredQ under 20 read 1'] < maxBadReads and \\\n",
    "        row['Number of bases with phredQ under 20 read 2'] < maxBadReads:\n",
    "        qualityTrimmed.append(index)\n",
    "        i += 1      \n",
    "dfQuality = dfRaw.filter(qualityTrimmed, axis = 'index')\n",
    "\n",
    "# Extract a new dataframe that only has rows from the quality dataframe with exactly one mutation and a barcode\n",
    "for index, row in dfQuality.iterrows():\n",
    "    if row['Number of mutations in read 1'] + row['Number of mutations in read 2'] == 1 and \\\n",
    "        row['Is there a barcode in read 1?'] ^ row['Is there a barcode in read 2?']:\n",
    "        mutantRows.append(index)\n",
    "        i += 1\n",
    "dfMutants = dfQuality.filter(mutantRows, axis = 'index')\n",
    "\n",
    "# In the single-mutants dataframe make a new column that contains the unique mutation\n",
    "dummyMutation = 'Mutation was not added'\n",
    "dfMutants['Single mutation'] = dummyMutation\n",
    "dfMutants['Single mutation'] = np.where(dfMutants['Mutation1'] == 'no mutation', \n",
    "                                        dfMutants['Mutation2'], \n",
    "                                        dfMutants['Mutation1']) \n",
    "\n",
    "# Make a dataframe that contains any row with a barcode and zero or 1 mutations\n",
    "for index, row in dfQuality.iterrows():\n",
    "    if row['Number of mutations in read 1'] + row['Number of mutations in read 2'] <= 1 and \\\n",
    "        row['Is there a barcode in read 1?'] ^ row['Is there a barcode in read 2?']:\n",
    "        WTorSingleRow.append(index)\n",
    "        i += 1      \n",
    "dfWTorSingle = dfQuality.filter(WTorSingleRow, axis = 'index')\n",
    "\n",
    "\n",
    "\n",
    "# How many of each mutant are there?\n",
    "\n",
    "print('Single mutant counts: ')\n",
    "dfMutants['Single mutation'].value_counts()\n",
    "\n",
    "# How many of each barcode are there?\n",
    "print('Single mutant barcode counts: ')\n",
    "dfMutants['Barcode'].value_counts()\n",
    "\n",
    "print('Total mutant barcode counts: ')\n",
    "dfWTorSingle['Barcode'].value_counts()\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "dfMutants.head()\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Single mutant counts: \n",
      "Y447N    61441\n",
      "E380Q      937\n",
      "F351L      887\n",
      "A385T      779\n",
      "G394R      709\n",
      "         ...  \n",
      "N381P        1\n",
      "R451E        1\n",
      "N384F        1\n",
      "G377Y        1\n",
      "F379K        1\n",
      "Name: Single mutation, Length: 1180, dtype: int64\n",
      "Single mutant barcode counts: \n",
      "ACTAGACCAGGGAAT    274\n",
      "CCAATGTGAGCAGCG    273\n",
      "CGACAAGCAGCCCTT    253\n",
      "CGCAGGTCCCACCGA    234\n",
      "CATCCGAGCTCGCAG    229\n",
      "                  ... \n",
      "TGCCGGGCTCCACAA      1\n",
      "GCAAAGCATCATAAA      1\n",
      "GCGAGAATATGTAGA      1\n",
      "ATGCGGCCCTTCCCT      1\n",
      "ATGCGAGCATGGGAA      1\n",
      "Name: Barcode, Length: 23531, dtype: int64\n",
      "Total mutant barcode counts: \n",
      "GGAGGGAACACTAGT    1087\n",
      "CATCCGAGCTCGCAG    1028\n",
      "TCTCGTGCACGTCCT     955\n",
      "TCTACATCTGGGATG     868\n",
      "ATACCAACTATTTAG     848\n",
      "                   ... \n",
      "GAAGTGAACGCGACA       1\n",
      "TCCTCGGGTGTGTGC       1\n",
      "ACTGGCTGTGGCCAA       1\n",
      "GCCTCTGAGCGCACT       1\n",
      "CGCCCTCACTCGTTC       1\n",
      "Name: Barcode, Length: 36434, dtype: int64\n",
      "Single mutation\n",
      "A361C      7\n",
      "A361D    162\n",
      "A361E     63\n",
      "A361F     47\n",
      "A361G     67\n",
      "Name: Barcode, dtype: int64\n",
      "Barcode\n",
      "AAAAAAAAACCGGCA     3\n",
      "AAAAAAAAACGAATG    12\n",
      "AAAAAAAAAGGCATT     2\n",
      "AAAAAAAAATGAGTA     3\n",
      "AAAAAAAACAAAAAT     1\n",
      "Name: Single mutation, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# How many of each mutant are there?\n",
    "print('Single mutant counts: ')\n",
    "print(str(dfMutants['Single mutation'].value_counts()))\n",
    "\n",
    "# How many of each barcode are there among single mutants?\n",
    "print('Single mutant barcode counts: ')\n",
    "print(dfMutants['Barcode'].value_counts())\n",
    "\n",
    "# How many of each barcode are there among among zero and single mutants?\n",
    "print('Total mutant barcode counts: ')\n",
    "print(str(dfWTorSingle['Barcode'].value_counts()))\n",
    "\n",
    "# How many barcodes are there for each mutation? Answer as a histogram\n",
    "print(str(dfMutants.groupby('Single mutation')['Barcode'].nunique().head()))\n",
    "\n",
    "# How many mutations are there for each barcode? Answer as a histogram\n",
    "print(str(dfMutants.groupby('Barcode')['Single mutation'].nunique().head()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([0.000e+00, 7.441e+03, 3.326e+03, 2.594e+03, 2.014e+03, 1.754e+03,\n",
       "        1.319e+03, 1.068e+03, 7.930e+02, 6.260e+02, 5.020e+02, 3.970e+02,\n",
       "        2.870e+02, 2.520e+02, 1.790e+02, 1.510e+02, 1.310e+02, 1.000e+02,\n",
       "        1.020e+02, 7.900e+01, 5.500e+01, 4.800e+01, 3.900e+01, 3.100e+01,\n",
       "        2.600e+01, 2.300e+01, 2.000e+01, 1.600e+01, 1.500e+01, 1.600e+01,\n",
       "        1.700e+01, 9.000e+00, 1.300e+01, 1.000e+01, 8.000e+00, 2.000e+00,\n",
       "        5.000e+00, 7.000e+00, 1.000e+00, 4.000e+00, 6.000e+00, 2.000e+00,\n",
       "        2.000e+00, 2.000e+00, 1.000e+00, 3.000e+00, 2.000e+00, 2.000e+00,\n",
       "        0.000e+00, 1.000e+00, 0.000e+00, 1.000e+00, 3.000e+00, 2.000e+00,\n",
       "        1.000e+00, 4.000e+00, 0.000e+00, 2.000e+00, 2.000e+00, 1.000e+00,\n",
       "        1.000e+00, 0.000e+00, 1.000e+00, 0.000e+00, 0.000e+00, 1.000e+00,\n",
       "        0.000e+00, 0.000e+00, 1.000e+00, 0.000e+00, 0.000e+00, 1.000e+00,\n",
       "        0.000e+00, 0.000e+00, 1.000e+00, 1.000e+00, 0.000e+00, 0.000e+00,\n",
       "        1.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 2.000e+00,\n",
       "        0.000e+00, 0.000e+00, 1.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,\n",
       "        0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,\n",
       "        1.000e+00, 0.000e+00, 0.000e+00]),\n",
       " array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,\n",
       "        17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,\n",
       "        34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,\n",
       "        51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,\n",
       "        68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,\n",
       "        85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]),\n",
       " <BarContainer object of 99 artists>)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAATW0lEQVR4nO3df4xd5X3n8fenEJKW7sYmzFqsbdasYiWiKwVYCxylqrqwMYZUMX+kiKgqI2TJ+wdtk1WlrrP7h1VoJCKtSoO0RULg1kTZEEqTxUpQWK9DVfUPCCawhB9hPSGhtgXYjQ1pg5qU9Lt/3GforTPD3Inv3LHneb+k0T3ne5577vNw0Oecee6Z41QVkqQ+/Nxyd0CSNDmGviR1xNCXpI4Y+pLUEUNfkjpy9nJ34O2cf/75tWHDhuXuhiSdUZ544om/qaqpubad1qG/YcMGDhw4sNzdkKQzSpKX5tvm9I4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXktP6L3HHasPOrby1/77aPLGNPJGn5eKUvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6siCoZ/kfUmeGvr5QZJPJjkvyb4kB9vr6tY+Se5IMpPk6SSXDe1rurU/mGR6KQcmSfppC4Z+Vb1QVZdU1SXAvwfeAL4M7AT2V9VGYH9bB7gG2Nh+dgB3AiQ5D9gFXAFcDuyaPVFIkiZjsdM7VwHfqaqXgG3AnlbfA1zXlrcB99bAo8CqJBcAVwP7qup4VZ0A9gFbT3UAkqTRLTb0bwC+0JbXVNXLbfkVYE1bXgscGnrP4Vabr/7PJNmR5ECSA8eOHVtk9yRJb2fk0E9yDvBR4M9O3lZVBdQ4OlRVd1XVpqraNDU1NY5dSpKaxVzpXwN8s6pebeuvtmkb2uvRVj8CrB9637pWm68uSZqQxYT+x/mnqR2AvcDsHTjTwIND9RvbXTybgdfbNNDDwJYkq9sXuFtaTZI0ISM9Tz/JucCHgf80VL4NuD/JduAl4PpWfwi4FphhcKfPTQBVdTzJrcDjrd0tVXX8lEcgSRrZSKFfVT8E3nNS7fsM7uY5uW0BN8+zn93A7sV3U5I0Dv5FriR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktSRkUI/yaokDyT5dpLnk3wwyXlJ9iU52F5Xt7ZJckeSmSRPJ7lsaD/Trf3BJNNLNShJ0txGvdL/LPC1qno/8AHgeWAnsL+qNgL72zrANcDG9rMDuBMgyXnALuAK4HJg1+yJQpI0GQuGfpJ3A78C3ANQVT+uqteAbcCe1mwPcF1b3gbcWwOPAquSXABcDeyrquNVdQLYB2wd41gkSQsY5Ur/IuAY8CdJnkxyd5JzgTVV9XJr8wqwpi2vBQ4Nvf9wq81X/2eS7EhyIMmBY8eOLW40kqS3NUronw1cBtxZVZcCP+SfpnIAqKoCahwdqqq7qmpTVW2ampoaxy4lSc0ooX8YOFxVj7X1BxicBF5t0za016Nt+xFg/dD717XafHVJ0oQsGPpV9QpwKMn7Wukq4DlgLzB7B8408GBb3gvc2O7i2Qy83qaBHga2JFndvsDd0mqSpAk5e8R2vw18Psk5wIvATQxOGPcn2Q68BFzf2j4EXAvMAG+0tlTV8SS3Ao+3drdU1fGxjEKSNJKRQr+qngI2zbHpqjnaFnDzPPvZDexeRP8kSWPkX+RKUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SerISKGf5HtJvpXkqSQHWu28JPuSHGyvq1s9Se5IMpPk6SSXDe1nurU/mGR6vs+TJC2NxVzp/4equqSqZv+t3J3A/qraCOxv6wDXABvbzw7gThicJIBdwBXA5cCu2ROFJGkyTmV6Zxuwpy3vAa4bqt9bA48Cq5JcAFwN7Kuq41V1AtgHbD2Fz5ckLdKooV/A/07yRJIdrbamql5uy68Aa9ryWuDQ0HsPt9p8dUnShJw9YrtfrqojSf4VsC/Jt4c3VlUlqXF0qJ1UdgBceOGF49ilJKkZ6Uq/qo6016PAlxnMyb/apm1or0db8yPA+qG3r2u1+eonf9ZdVbWpqjZNTU0tbjSSpLe1YOgnOTfJv5hdBrYAzwB7gdk7cKaBB9vyXuDGdhfPZuD1Ng30MLAlyer2Be6WVpMkTcgo0ztrgC8nmW3/P6vqa0keB+5Psh14Cbi+tX8IuBaYAd4AbgKoquNJbgUeb+1uqarjYxuJJGlBC4Z+Vb0IfGCO+veBq+aoF3DzPPvaDexefDclSePgX+RKUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHRk59JOcleTJJF9p6xcleSzJTJIvJjmn1d/Z1mfa9g1D+/hUq7+Q5Oqxj0aS9LYWc6X/CeD5ofXPALdX1XuBE8D2Vt8OnGj121s7klwM3AD8ErAV+OMkZ51a9yVJizFS6CdZB3wEuLutB7gSeKA12QNc15a3tXXa9qta+23AfVX1o6r6LjADXD6GMUiSRjTqlf4fAb8H/GNbfw/wWlW92dYPA2vb8lrgEEDb/npr/1Z9jve8JcmOJAeSHDh27NjoI5EkLWjB0E/ya8DRqnpiAv2hqu6qqk1VtWlqamoSHylJ3Th7hDYfAj6a5FrgXcC/BD4LrEpydruaXwccae2PAOuBw0nOBt4NfH+oPmv4PZKkCVjwSr+qPlVV66pqA4MvYr9eVb8BPAJ8rDWbBh5sy3vbOm3716uqWv2GdnfPRcBG4BtjG4kkaUGjXOnP578A9yX5A+BJ4J5Wvwf4XJIZ4DiDEwVV9WyS+4HngDeBm6vqJ6fw+ZKkRVpU6FfVXwB/0ZZfZI67b6rq74Ffn+f9nwY+vdhOSpLGw7/IlaSOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6sipPFr5jLVh51ffWv7ebR9Zxp5I0mR5pS9JHTH0Jakjhr4kdcTQl6SOGPqS1JEFQz/Ju5J8I8n/TfJskt9v9YuSPJZkJskXk5zT6u9s6zNt+4ahfX2q1V9IcvWSjUqSNKdRrvR/BFxZVR8ALgG2JtkMfAa4vareC5wAtrf224ETrX57a0eSi4EbgF8CtgJ/nOSsMY5FkrSABUO/Bv6urb6j/RRwJfBAq+8BrmvL29o6bftVSdLq91XVj6rqu8AMcPk4BiFJGs1Ic/pJzkryFHAU2Ad8B3itqt5sTQ4Da9vyWuAQQNv+OvCe4foc7xn+rB1JDiQ5cOzYsUUPSJI0v5FCv6p+UlWXAOsYXJ2/f6k6VFV3VdWmqto0NTW1VB8jSV1a1N07VfUa8AjwQWBVktnHOKwDjrTlI8B6gLb93cD3h+tzvEeSNAGj3L0zlWRVW/554MPA8wzC/2Ot2TTwYFve29Zp279eVdXqN7S7ey4CNgLfGNM4JEkjGOWBaxcAe9qdNj8H3F9VX0nyHHBfkj8AngTuae3vAT6XZAY4zuCOHarq2ST3A88BbwI3V9VPxjscSdLbWTD0q+pp4NI56i8yx903VfX3wK/Ps69PA59efDclSePgX+RKUkcMfUnqiKEvSR0x9CWpI4a+JHXE0JekjnT5D6MP8x9Jl9QTr/QlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1JEFQz/J+iSPJHkuybNJPtHq5yXZl+Rge13d6klyR5KZJE8nuWxoX9Ot/cEk00s3LEnSXEa50n8T+N2quhjYDNyc5GJgJ7C/qjYC+9s6wDXAxvazA7gTBicJYBdwBYN/UH3X7IlCkjQZC4Z+Vb1cVd9sy38LPA+sBbYBe1qzPcB1bXkbcG8NPAqsSnIBcDWwr6qOV9UJYB+wdZyDkSS9vUU9Tz/JBuBS4DFgTVW93Da9Aqxpy2uBQ0NvO9xq89VP/owdDH5D4MILL1xM906Zz9aXtNKN/EVukl8E/hz4ZFX9YHhbVRVQ4+hQVd1VVZuqatPU1NQ4dilJakYK/STvYBD4n6+qL7Xyq23ahvZ6tNWPAOuH3r6u1earS5ImZJS7dwLcAzxfVX84tGkvMHsHzjTw4FD9xnYXz2bg9TYN9DCwJcnq9gXullaTJE3IKHP6HwJ+E/hWkqda7b8CtwH3J9kOvARc37Y9BFwLzABvADcBVNXxJLcCj7d2t1TV8XEMQpI0mgVDv6r+Csg8m6+ao30BN8+zr93A7sV0UJI0Pv5FriR1ZFG3bPbE2zclrURe6UtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR3x2Tsj8Dk8klYKr/QlqSOGviR1xNCXpI4Y+pLUEUNfkjqyYOgn2Z3kaJJnhmrnJdmX5GB7Xd3qSXJHkpkkTye5bOg90639wSTTSzMcSdLbGeVK/0+BrSfVdgL7q2ojsL+tA1wDbGw/O4A7YXCSAHYBVwCXA7tmTxRnmg07v/rWjySdaRYM/ar6S+D4SeVtwJ62vAe4bqh+bw08CqxKcgFwNbCvqo5X1QlgHz99IpEkLbGfdU5/TVW93JZfAda05bXAoaF2h1ttvvpPSbIjyYEkB44dO/Yzdk+SNJdT/iK3qgqoMfRldn93VdWmqto0NTU1rt1KkvjZQ//VNm1Dez3a6keA9UPt1rXafHVJ0gT9rKG/F5i9A2caeHCofmO7i2cz8HqbBnoY2JJkdfsCd0urSZImaMEHriX5AvCrwPlJDjO4C+c24P4k24GXgOtb84eAa4EZ4A3gJoCqOp7kVuDx1u6Wqjr5y2FJ0hJbMPSr6uPzbLpqjrYF3DzPfnYDuxfVO0nSWPlo5VPgI5clnWl8DIMkdcTQl6SOGPqS1BHn9MfE+X1JZwKv9CWpI4a+JHXE0JekjjinvwSc35d0uvJKX5I6YuhLUkec3lliTvVIOp14pS9JHTH0JakjTu9M0PBUDzjdI2nyDP1l5Hy/pElzekeSOuKV/mnCq35Jk2Don4ZOnvuf5clA0qmaeOgn2Qp8FjgLuLuqbpt0H85UngwknaqJhn6Ss4D/AXwYOAw8nmRvVT03yX6sNPOdDIZ5YpAEk7/SvxyYqaoXAZLcB2wDDP0lNsqJYZyGTzJ+XyGdPiYd+muBQ0Prh4Erhhsk2QHsaKt/l+SFU/i884G/OYX3n4lOizHnM4urj8FpMe4Jc8z9WOy4/818G067L3Kr6i7grnHsK8mBqto0jn2dKXocM/Q5bsfcj3GOe9L36R8B1g+tr2s1SdIETDr0Hwc2JrkoyTnADcDeCfdBkro10emdqnozyW8BDzO4ZXN3VT27hB85lmmiM0yPY4Y+x+2Y+zG2caeqxrUvSdJpzmfvSFJHDH1J6siKDP0kW5O8kGQmyc7l7s9SSLI+ySNJnkvybJJPtPp5SfYlOdheVy93X5dCkrOSPJnkK239oiSPtWP+xXajwIqRZFWSB5J8O8nzST7Yw7FO8p/b/9/PJPlCknetxGOdZHeSo0meGarNeXwzcEcb/9NJLlvMZ6240B961MM1wMXAx5NcvLy9WhJvAr9bVRcDm4Gb2zh3AvuraiOwv62vRJ8Anh9a/wxwe1W9FzgBbF+WXi2dzwJfq6r3Ax9gMPYVfayTrAV+B9hUVf+Owc0fN7Ayj/WfAltPqs13fK8BNrafHcCdi/mgFRf6DD3qoap+DMw+6mFFqaqXq+qbbflvGYTAWgZj3dOa7QGuW5YOLqEk64CPAHe39QBXAg+0Jitq3EneDfwKcA9AVf24ql6jg2PN4A7Dn09yNvALwMuswGNdVX8JHD+pPN/x3QbcWwOPAquSXDDqZ63E0J/rUQ9rl6kvE5FkA3Ap8BiwpqpebpteAdYsV7+W0B8Bvwf8Y1t/D/BaVb3Z1lfaMb8IOAb8SZvSujvJuazwY11VR4D/Dvw1g7B/HXiClX2sh813fE8p41Zi6HclyS8Cfw58sqp+MLytBvfjrqh7cpP8GnC0qp5Y7r5M0NnAZcCdVXUp8ENOmspZocd6NYOr2ouAfw2cy09PgXRhnMd3JYZ+N496SPIOBoH/+ar6Uiu/OvurXns9ulz9WyIfAj6a5HsMpu6uZDDfvapNAcDKO+aHgcNV9Vhbf4DBSWClH+v/CHy3qo5V1T8AX2Jw/FfysR423/E9pYxbiaHfxaMe2jz2PcDzVfWHQ5v2AtNteRp4cNJ9W0pV9amqWldVGxgc269X1W8AjwAfa81W1Lir6hXgUJL3tdJVDB5HvqKPNYNpnc1JfqH9/z477hV7rE8y3/HdC9zY7uLZDLw+NA20sKpacT/AtcD/A74D/Lfl7s8SjfGXGfy69zTwVPu5lsH89n7gIPB/gPOWu69L+N/gV4GvtOV/C3wDmAH+DHjncvdvzGO9BDjQjvf/Alb3cKyB3we+DTwDfA5450o81sAXGHxv8Q8MfrPbPt/xBcLgDsXvAN9icHfTyJ/lYxgkqSMrcXpHkjQPQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR15P8DXO+WH9VcpkgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "MutsperBarc = dfMutants.groupby('Barcode')['Single mutation'].nunique()\n",
    "plt.hist(MutsperBarc, bins = range(0,100,1))\n",
    "#print('This is a histogram of how many unique mutations each barcode has')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is a histogram of how many unique barcodes each mutation has\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAD4CAYAAAD1jb0+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAPeUlEQVR4nO3dX4xc9XnG8eepbSAKqJh6ZFkGdQ1FiXzRGGvlEgWhFEpqfGOQUAUX1BdUjlqQQEov3ERqidSLpCogVYqIjLDiVpQ/5Y9AJW3iUksoUmu6JsbYWBRDHRXLeJdSArlJa3h7Mb+F6TBn58y/nXlnvx9ptWd+58yc97fHPMyceeeMI0IAgHx+ZdwFAAD6Q4ADQFIEOAAkRYADQFIEOAAktXo5d7Zu3bqYmZlZzl0CQHqHDx9+NyIa7ePLGuAzMzOam5tbzl0CQHq2f9ZpnFMoAJAUAQ4ASRHgAJAUAQ4ASRHgAJAUAQ4ASXUNcNsX2H7J9iu2j9v+dhnfZPuQ7ZO2H7d93ujLBQAsqvMM/JeSrouIL0naImm77aslfVfSAxHxG5L+W9IdI6sSAPAZXQM8mn5Rbq4pPyHpOklPlvH9km4aRYEAgM5qnQO3vcr2EUnzkg5IelPS+xFxrmzytqSNFffdbXvO9tzCwsJAxc7seX6g+wPANKkV4BHxUURskXSppG2Svlh3BxGxNyJmI2K20fjMR/kBAH3qqQslIt6XdFDSlyVdbHvxWiqXSjo93NIAAEup04XSsH1xWf6cpBsknVAzyG8pm+2S9OyIagQAdFDnaoQbJO23vUrNwH8iIv7e9muSHrP955J+KunhEdYJAGjTNcAj4qikqzqMv6Xm+XAAwBjwSUwASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASIoAB4CkCHAASKprgNu+zPZB26/ZPm777jJ+r+3Tto+Unx2jLxcAsGh1jW3OSfpGRLxs+yJJh20fKOseiIi/HF15AIAqXQM8Is5IOlOWP7R9QtLGURcGAFhaT+fAbc9IukrSoTJ0l+2jtvfZXltxn92252zPLSwsDFYtAOATtQPc9oWSnpJ0T0R8IOlBSVdI2qLmM/T7Ot0vIvZGxGxEzDYajcErBgBIqhngtteoGd6PRMTTkhQRZyPio4j4WNJDkraNrkwAQLs6XSiW9LCkExFxf8v4hpbNbpZ0bPjlAQCq1OlC+Yqk2yW9avtIGfumpNtsb5EUkk5J+voI6gMAVKjThfITSe6w6ofDLwcAUBefxASApAhwAEiKAAeApAhwAEiKAAeApAhwAEiKAAeApAhwAEiKAAeApAhwAEiKAAeApAhwAEiKAAeApAhwAEiKAAeApNIF+Mye58ddAgBMhHQBDgBoIsABICkCHACSIsABICkCHACSIsABICkCHACSIsABICkCHACSIsABIKmuAW77MtsHbb9m+7jtu8v4JbYP2H6j/F47+nIBAIvqPAM/J+kbEbFZ0tWS7rS9WdIeSS9ExJWSXii3AQDLpGuAR8SZiHi5LH8o6YSkjZJ2StpfNtsv6aYR1QgA6KCnc+C2ZyRdJemQpPURcaasekfS+or77LY9Z3tuYWFhkFoBAC1qB7jtCyU9JemeiPigdV1EhKTodL+I2BsRsxEx22g0BioWAPCpWgFue42a4f1IRDxdhs/a3lDWb5A0P5oSAQCd1OlCsaSHJZ2IiPtbVj0naVdZ3iXp2eGXBwCosrrGNl+RdLukV20fKWPflPQdSU/YvkPSzyT93kgqBAB01DXAI+Inklyx+vrhlgMAqItPYgJAUlMb4Hz5MYBpN7UBDgDTjgAHgKQIcABIigAHgKQIcABIigAHgKRSB/jMnudpFwSwYqUOcABYyQhwAEiKAAeApAhwAEiKAAeApKYqwOlIAbCSTFWAA8BKQoADQFIEOAAkRYADQFIEOAAkRYADQFJTHeC9thVOSxsiF/kCVoapDnAAmGYEOAAkRYADQFJdA9z2Ptvzto+1jN1r+7TtI+Vnx2jLBAC0q/MM/AeStncYfyAitpSfHw63LABAN10DPCJelPTeMtQCAOjBIOfA77J9tJxiWVu1ke3dtudszy0sLAywu8Et1VpXtY52PACTqt8Af1DSFZK2SDoj6b6qDSNib0TMRsRso9Hoc3cAgHZ9BXhEnI2IjyLiY0kPSdo23LIAAN30FeC2N7TcvFnSsaptAQCjsbrbBrYflfRVSetsvy3pzyR91fYWSSHplKSvj65EAEAnXQM8Im7rMPzwCGoBAPSAT2ICQFIpA3wcVxmknRDApEkZ4AAAAhwA0iLAASApAhwAkiLAASApAhwAkpqKAG//Et86LX+DtAX2+6XBfNkwgGGaigAHgJWIAAeApAhwAEiKAAeApAhwAEiKAAeApKYuwIdxpcJ+W/3a70fLIIBRmroAB4CVggAHgKQIcABIigAHgKQIcABIigAHgKRWXIC3XhGwbgvhcrYD9nuVQwArz4oLcACYFgQ4ACRFgANAUl0D3PY+2/O2j7WMXWL7gO03yu+1oy0TANCuzjPwH0ja3ja2R9ILEXGlpBfKbQDAMuoa4BHxoqT32oZ3StpflvdLumm4ZQEAuun3HPj6iDhTlt+RtL5qQ9u7bc/ZnltYWOhzd6NV1YZXp6VwqZbEUdQEAIsGfhMzIkJSLLF+b0TMRsRso9EYdHcAgKLfAD9re4Mkld/zwysJAFBHvwH+nKRdZXmXpGeHUw4AoK46bYSPSvoXSV+w/bbtOyR9R9INtt+Q9DvlNgBgGa3utkFE3Fax6voh1wIA6AGfxASApNIGeN02u2G2+dV5jPb9dbvyYbfHHPfVEQFMrrQBDgArHQEOAEkR4ACQFAEOAEkR4ACQFAEOAEkR4AOo26I4rPWt29XZdy9tj/2gnREYLwIcAJIiwAEgKQIcAJIiwAEgKQIcAJIiwCtMQocFNXxqUuoAJgkBDgBJEeAAkBQBDgBJEeAAkBQBDgBJEeAAkBQBPgJ1Lk7Vvs2wvrOT79DsjL8BphEBDgBJEeAAkBQBDgBJrR7kzrZPSfpQ0keSzkXE7DCKAgB0N1CAF78dEe8O4XEAAD3gFAoAJDVogIekH9s+bHt3pw1s77Y9Z3tuYWFhwN1Nt1G0utX93s5Rqbvfftofxzkn2hIxCQYN8GsiYqukGyXdafva9g0iYm9EzEbEbKPRGHB3AIBFAwV4RJwuv+clPSNp2zCKAgB013eA2/687YsWlyV9TdKxYRUGAFjaIF0o6yU9Y3vxcf42Iv5xKFUBALrqO8Aj4i1JXxpiLQCAHtBGCABJEeDJtLav9drKttj+1utVC5faTy81tLY0jru9cdiWex7Lsb9pOTbTjAAHgKQIcABIigAHgKQIcABIigAHgKQIcABIigBPYhRX5luqJXGplsKqL2UexlX6hj3PQa6G2Do+6N93kG2Xq52vbmsp7YVNk/B3IMABICkCHACSIsABICkCHACSIsABICkCHACSIsAn0DivbNfrlQq7rauzbbfHbm/lW6qNcal91rmSY6crNvbTYljn71V1v16vGFn1eHX+Jp226/cql8MwCa15iyaplioEOAAkRYADQFIEOAAkRYADQFIEOAAkRYADQFIEOGrr5cp+7W14o2pNrLqKYt22wl7a5/q5Mt9SLX69PH5V62Qv++vUjtlLDZ3G66yre/+686vbFtqpVbLuv7l+tuu0PKyrdFYhwAEgKQIcAJIiwAEgqYEC3PZ226/bPml7z7CKAgB013eA214l6XuSbpS0WdJttjcPqzAAwNIGeQa+TdLJiHgrIv5H0mOSdg6nLABAN46I/u5o3yJpe0T8Qbl9u6Tfioi72rbbLWl3ufkFSa/3Wes6Se/2ed9Jw1wmE3OZTMxF+vWIaLQPrh68nqVFxF5Jewd9HNtzETE7hJLGjrlMJuYymZhLtUFOoZyWdFnL7UvLGABgGQwS4P8m6Urbm2yfJ+lWSc8NpywAQDd9n0KJiHO275L0I0mrJO2LiONDq+yzBj4NM0GYy2RiLpOJuVTo+01MAMB48UlMAEiKAAeApFIEePaP7Ns+ZftV20dsz5WxS2wfsP1G+b123HV2Ynuf7Xnbx1rGOtbupr8qx+mo7a3jq/z/q5jHvbZPl+NyxPaOlnV/Uubxuu3fHU/Vndm+zPZB26/ZPm777jKe8bhUzSXdsbF9ge2XbL9S5vLtMr7J9qFS8+Ol6UO2zy+3T5b1Mz3vNCIm+kfNN0jflHS5pPMkvSJp87jr6nEOpyStaxv7C0l7yvIeSd8dd50VtV8raaukY91ql7RD0j9IsqSrJR0ad/1d5nGvpD/usO3m8u/sfEmbyr+/VeOeQ0t9GyRtLcsXSfr3UnPG41I1l3THpvx9LyzLayQdKn/vJyTdWsa/L+kPy/IfSfp+Wb5V0uO97jPDM/Bp/cj+Tkn7y/J+STeNr5RqEfGipPfahqtq3ynpr6PpXyVdbHvDshTaRcU8quyU9FhE/DIi/kPSSTX/HU6EiDgTES+X5Q8lnZC0UTmPS9VcqkzssSl/31+Um2vKT0i6TtKTZbz9uCweryclXW/bvewzQ4BvlPSfLbff1tIHeBKFpB/bPlwuLSBJ6yPiTFl+R9L68ZTWl6raMx6ru8pphX0tp7HSzKO87L5KzWd7qY9L21ykhMfG9irbRyTNSzqg5iuE9yPiXNmktd5P5lLW/1zSr/WyvwwBPg2uiYital658U7b17aujOZrqJT9nJlrl/SgpCskbZF0RtJ9Y62mR7YvlPSUpHsi4oPWddmOS4e5pDw2EfFRRGxR85Pp2yR9cZT7yxDg6T+yHxGny+95Sc+oeWDPLr6MLb/nx1dhz6pqT3WsIuJs+Q/uY0kP6dOX4hM/D9tr1Ay8RyLi6TKc8rh0mkvmYyNJEfG+pIOSvqzmKavFD0221vvJXMr6X5X0X73sJ0OAp/7Ivu3P275ocVnS1yQdU3MOu8pmuyQ9O54K+1JV+3OSfr90PVwt6ectL+knTtt54JvVPC5Scx63li6BTZKulPTSctdXpZwnfVjSiYi4v2VVuuNSNZeMx8Z2w/bFZflzkm5Q85z+QUm3lM3aj8vi8bpF0j+XV071jfud25rv7u5Q893pNyV9a9z19Fj75Wq+a/6KpOOL9at5rusFSW9I+idJl4y71or6H1XzJez/qnn+7o6q2tV8F/575Ti9Kml23PV3mcfflDqPlv+YNrRs/60yj9cl3Tju+tvmco2ap0eOSjpSfnYkPS5Vc0l3bCT9pqSflpqPSfrTMn65mv+TOSnp7ySdX8YvKLdPlvWX97pPPkoPAEllOIUCAOiAAAeApAhwAEiKAAeApAhwAEiKAAeApAhwAEjq/wBx6K11nRoIWgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "BarcsperMut = dfMutants.groupby('Single mutation')['Barcode'].nunique()\n",
    "plt.hist(BarcsperMut, bins = range(0,300,1))\n",
    "\n",
    "\n",
    "\n",
    "print('This is a histogram of how many unique barcodes each mutation has')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAPd0lEQVR4nO3dX6zkZ13H8ffHrgUtWoSuhOwWt7ib6l4omJMCgQskYrbAtgaJdkMimk03NdZgYqLbaEy8Um5ESSq4Cc3ekNaKqNuypiB/0psGegoFdllXDk1Jd4PugUJNjLEWvl6c37bDeP785szMzpln3q/kZM8885v5Pc9h+Jyn399znl+qCklSW35o1h2QJE2e4S5JDTLcJalBhrskNchwl6QG7Zp1BwCuu+662rdv36y7IUlz5bHHHvtWVe1e77kdEe779u1jeXl51t2QpLmS5BsbPTfTskySw0lOPPPMM7PshiQ1Z6bhXlUPVNWxa6+9dpbdkKTmOHOXpAY5c5ekBrkUUpIaZLhLUoOsuUtSg6y5S1KDLMtIUoMMd0lqkDV3SWqQNXdJapBlGUlqkOEuSQ0y3CWpQV5QlaQGeUFVkhpkWUaSGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHXuktQg17lLUoMsy0hSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1KCphHuSa5IsJ3nHNN5fkrS5XuGe5J4kl5KcGWo/lOR8kpUkxwee+kPg/kl2VJLUX9+Z+0ng0GBDkquAu4GbgYPAkSQHk7wV+CpwaYL9lCSNYFefg6rq4ST7hppvAlaq6gmAJPcBtwIvAa5hLfD/O8npqvr+5LosSdpKr3DfwB7gqYHHF4DXVdWdAEl+E/jWRsGe5BhwDOBVr3rVGN2QJA2b2mqZqjpZVQ9u8vyJqlqqqqXdu3dPqxuStJDGCfeLwPUDj/d2bb255a8kTcc44f4ocCDJDUmuBm4DTo3yBm75K0nT0Xcp5L3AI8CNSS4kOVpVzwF3Ag8B54D7q+rsKCd35i5J05GqmnUfWFpaquXl5Vl3Q5LmSpLHqmppvee8zZ4kNcjb7ElSg9w4TJIaZFlGkhpkWUaSGmRZRpIaZFlGkhpkWUaSGmRZRpIaZLhLUoMMd0lqkBdUJalBXlCVpAZZlpGkBjUR7vuOf3zWXZCkHaWJcJck/SAvqEpSg7ygKkkNsiwjSQ0y3CWpQYa7JDXIcJekBhnuktQgl0JKUoNcCilJDbIsI0kNMtwlqUGGuyQ1yHCXpAY1E+5u+ytJL2gm3CVJLzDcJalBhrskNWji4Z7kZ5N8KMlHk/z2pN9fkrS1XuGe5J4kl5KcGWo/lOR8kpUkxwGq6lxV3QH8GvDGyXdZkrSVvjP3k8ChwYYkVwF3AzcDB4EjSQ52z90CfBw4PbGeSpJ66xXuVfUw8PRQ803ASlU9UVXPAvcBt3bHn6qqm4F3b/SeSY4lWU6yvLq6ur3eS5LWtWuM1+4Bnhp4fAF4XZI3A+8EXsQmM/eqOgGcAFhaWqox+iFJGjJOuK+rqj4LfLbPsUkOA4f3798/6W5I0kIbZ7XMReD6gcd7u7be3PJXkqZjnHB/FDiQ5IYkVwO3AadGeYNJ36zDLQgkaU3fpZD3Ao8ANya5kORoVT0H3Ak8BJwD7q+qs6Oc3Jm7JE1Hr5p7VR3ZoP00LneUpB3He6hKUoOau4eqdXdJcuMwSWqSZRlJalBzZRlJkmUZSWpSk2UZL6pKWnSWZSSpQU2XZZzBS1pUTYe7JC2qJmvu4Kxd0mKz5i5JDWq+LOMMXtIiaj7cJWkRLUy4O4OXtEiavaC6GYNeUusmfoPsUVTVA8ADS0tLt0/zPJfD3FCXtCgWpiwzzKCX1LKFDXdYC3hDXlKLFjrcJalVhjub1+Sd2UuaR4Z7ZzDgDXRJ885w78GwlzRvFnKdex/DgW7AS5onbhy2CQNd0ryyLCNJDTLcR+DFVknzwnDfhuGQN/Al7TSG+xgMeEk7leEuSQ0y3Me03pJJZ/GSZs1wnyDLNJJ2iqns557kV4C3Az8OfLiqPjGN80iS1td75p7kniSXkpwZaj+U5HySlSTHAarqH6vqduAO4Ncn2+X54exd0qyMUpY5CRwabEhyFXA3cDNwEDiS5ODAIX/cPb+wBmvwhr2kK6V3uFfVw8DTQ803AStV9URVPQvcB9yaNe8D/rmqvjC57kqS+hj3guoe4KmBxxe6tt8Ffgl4V5I71nthkmNJlpMsr66ujtmNnc9Zu6QraSoXVKvqA8AHtjjmBHACYGlpqabRD0laVOPO3C8C1w883tu19bKTt/y9UpzRS5qGccP9UeBAkhuSXA3cBpzq++KdvuXvNGx0cdWQlzRJoyyFvBd4BLgxyYUkR6vqOeBO4CHgHHB/VZ0d4T0Xcubu6hlJ09a75l5VRzZoPw2c3s7Jq+oB4IGlpaXbt/N6SdL6vM3eDrDRDN6ZvaTt8jZ7O8h6m44Z8JK2w43DdiADXtK4LMtIUoMsy+xQztYljcOyjCQ1yHCfI87mJfVlzX1OGOySRmHNfQ5ttZrGXwSSLMvMmY3u02qgSxpkuEtSg6y5zzln7JLWY829UYa+tNgsyzTGOrwkMNwlqUmG+4JYb8dJSe3yguqcMqglbcYLqgvOXxJSmyzLNGS9oO5TjvEvXKX2GO4LzBCX2tX7Btlqw2b3a33yz99+hXsjaVqcuUtSgwx3jcRSjjQfDHdtyCCX5pfr3PW87f6hk78EpJ3Hde4ai8Eu7UyWZSSpQYa7rghn+NKVZbhrZAa1tPMZ7trSehdaN9rqQNLOYLhLUoMMd/XWd2a+2RYH47yvpP4Md0lq0MTDPcmrk3w4yUcn/d7aWcaZcTtbl6arV7gnuSfJpSRnhtoPJTmfZCXJcYCqeqKqjk6js5KkfvrO3E8ChwYbklwF3A3cDBwEjiQ5ONHeaWYuz6zHrbNPg7N+aWu9wr2qHgaeHmq+CVjpZurPAvcBt/Y9cZJjSZaTLK+urvbusOZDnwCeVUj7y0GLYJya+x7gqYHHF4A9SV6e5EPAa5PctdGLq+pEVS1V1dLu3bvH6IYkadjEL6hW1ber6o6q+umq+rPNjnVXyPaNcv/WrUpBzril/sYJ94vA9QOP93ZtvbkrpCRNxzjh/ihwIMkNSa4GbgNOjfIGztzb02cm3rcev9X+8uudazv9lFrUdynkvcAjwI1JLiQ5WlXPAXcCDwHngPur6uwoJ3fmLknTsavPQVV1ZIP208DpifZIkjQ2b7OnK2qUC6yjtl1ud8dKydvsSVKT3DhMkhrUq+Y+LUkOA4f3798/y25oDoyz9t2SjBaRZRlJapBlGUlqkKtl1LRRdrXc6o+ttvqjqu2cV5oWyzKS1CDLMpLUIMNdkhpkzV07wiRq1MMbiY37166jHrOdYwdfs516/rg/N68NtMuauyQ1yLKMJDXIcJekBhnuktQgw12SGuRqGS2scVaKbLQyZ9I39x5lFc2453XlTFtcLSNJDbIsI0kNMtwlqUGGuyQ1yHCXpAYZ7pLUIJdCaq5MelngODfk6HO+Pu8/zjLJWb5+o/ebxLE7aVnmdjeCmzWXQkpSgyzLSFKDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhq0a9JvmOQa4K+BZ4HPVtVHJn0OSdLmes3ck9yT5FKSM0Pth5KcT7KS5HjX/E7go1V1O3DLhPsrSeqhb1nmJHBosCHJVcDdwM3AQeBIkoPAXuCp7rDvTaabkqRR9Ar3qnoYeHqo+SZgpaqeqKpngfuAW4ELrAX8pu+f5FiS5STLq6uro/dcmqLBTbU22uhr+LlRNsPa7F6rm2041uccfTYs67N52fA4Rx3r8Ou2+ne977fa7G29c/S5p+1W/R4+13ba+p5vWsa5oLqHF2bosBbqe4CPAb+a5IPAAxu9uKpOVNVSVS3t3r17jG5IkoZN/IJqVf0X8Ft9jk1yGDi8f//+SXdDkhbaODP3i8D1A4/3dm29ueWvJE3HOOH+KHAgyQ1JrgZuA06N8gberEOSpqPvUsh7gUeAG5NcSHK0qp4D7gQeAs4B91fV2VFO7sxdkqajV829qo5s0H4aOD3RHkmSxuY9VCWpQd5DVZIa5MxdkhqUqpp1H0iyCnxjmy+/DvjWBLszDxzzYnDMi2GcMf9UVa37V6A7ItzHkWS5qpZm3Y8ryTEvBse8GKY1Zvdzl6QGGe6S1KAWwv3ErDswA455MTjmxTCVMc99zV2S9P+1MHOXJA0x3CWpQXMd7hvcw3XurXfP2iQvS/LJJF/r/v2Jrj1JPtD9DL6c5Bdm1/PtS3J9ks8k+WqSs0ne27U3O+4kL07y+SRf6sb8p137DUk+143tb7tdV0nyou7xSvf8vpkOYJuSXJXki0ke7B43PV6AJE8m+UqSx5Msd21T/WzPbbhvcg/XFpxk6J61wHHgU1V1APhU9xjWxn+g+zoGfPAK9XHSngN+v6oOAq8Hfqf737Plcf8P8Jaq+nngNcChJK8H3ge8v6r2A98BjnbHHwW+07W/vztuHr2XtZ1kL2t9vJf9YlW9ZmBN+3Q/21U1l1/AG4CHBh7fBdw1635NcHz7gDMDj88Dr+y+fyVwvvv+b4Aj6x03z1/APwFvXZRxAz8KfAF4HWt/rbira3/+c87a9tpv6L7f1R2XWfd9xHHu7YLsLcCDQFoe78C4nwSuG2qb6md7bmfubHwP11a9oqq+2X3/78Aruu+b+zl0//n9WuBzND7urkTxOHAJ+CTwdeC7tXa/BPjBcT0/5u75Z4CXX9EOj+8vgT8Avt89fjltj/eyAj6R5LEkx7q2qX62J34PVU1fVVWSJtewJnkJ8PfA71XVfyZ5/rkWx11V3wNek+SlwD8APzPbHk1PkncAl6rqsSRvnnF3rrQ3VdXFJD8JfDLJvw4+OY3P9jzP3Me+h+uc+Y8krwTo/r3UtTfzc0jyw6wF+0eq6mNdc/PjBqiq7wKfYa0s8dIklydeg+N6fszd89cC376yPR3LG4FbkjwJ3MdaaeavaHe8z6uqi92/l1j7JX4TU/5sz3O4j30P1zlzCnhP9/17WKtJX27/je4K++uBZwb+U29uZG2K/mHgXFX9xcBTzY47ye5uxk6SH2HtGsM51kL+Xd1hw2O+/LN4F/Dp6oqy86Cq7qqqvVW1j7X/v366qt5No+O9LMk1SX7s8vfALwNnmPZne9YXGsa8SPE24N9Yq1P+0az7M8Fx3Qt8E/hf1uptR1mrNX4K+BrwL8DLumPD2qqhrwNfAZZm3f9tjvlNrNUlvww83n29reVxAz8HfLEb8xngT7r2VwOfB1aAvwNe1LW/uHu80j3/6lmPYYyxvxl4cBHG243vS93X2ctZNe3PttsPSFKD5rksI0nagOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGvR/7Ytxr3i9mEIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# plt.hist(dfMutants.groupby('Barcode')['Single mutation'].nunique(), density=True, bins=10)\n",
    "plt.hist(dfRaw['Barcode'].value_counts(), bins = range(0,500,1))\n",
    "plt.yscale('log')\n",
    "\n",
    "# plt.hist(dfMutants.groupby('Single mutation')['Barcode'].nunique(), density=True, bins=10)\n",
    "# plt.hist(dfMutants['Single mutation'].value_counts(), density=True, bins=10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "parse() missing 2 required positional arguments: 'handle' and 'format'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-12dd2093f6e2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mrecords\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mrec\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mlines\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrec\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mletter_annotations\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m20\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: parse() missing 2 required positional arguments: 'handle' and 'format'"
     ]
    }
   ],
   "source": [
    "#this should allow for filtering reads by quality\n",
    "\n",
    "records = []\n",
    "for rec in SeqIO.parse():\n",
    "    lines += 1\n",
    "    if min(rec.letter_annotations) > 20:\n",
    "        records.append(rec)\n",
    "        all_reads = [rec for rec in SeqIO.parse()]\n",
    "\n",
    "\n",
    "good_reads = (\n",
    "    rec\n",
    "    for rec in SeqIO.parse(\"SRR020192.fastq\", \"fastq\")\n",
    "    if min(rec.letter_annotations[\"phred_quality\"]) >= 20\n",
    ")\n",
    "count = SeqIO.write(good_reads, \"good_quality.fastq\", \"fastq\")\n",
    "print(\"Saved %i reads\" % count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This takes two files with paired reads and makes one file where the reads go one after the other\n",
    "#stolen from (http://bytesizebio.net/2011/11/10/short-bioinformatics-hacks-reading-mate-pairs-from-a-fastq-file/)\n",
    "\n",
    "from Bio import SeqIO\n",
    "import itertools\n",
    "import sys\n",
    "import os\n",
    "def merge_fastq(fastq_path1, fastq_path2, outpath):\n",
    "    outfile = open(outpath,\"w\")\n",
    "    fastq_iter1 = SeqIO.parse(open(fastq_path1),\"fastq\")\n",
    "    fastq_iter2 = SeqIO.parse(open(fastq_path2),\"fastq\")\n",
    "    for rec1, rec2 in itertools.zip_longest(fastq_iter1, fastq_iter2):\n",
    "        SeqIO.write([rec1,rec2], outfile, \"fastq\")\n",
    "    outfile.close()\n",
    "\n",
    "    \n",
    "#Put files to be merged here and the file you want to make\n",
    "    \n",
    "merge_fastq('NP-11-19-6_S3_L001_R1_001.fastq', 'NP-11-19-6_S3_L001_R2_001.fastq', 'NP-11-19-6-merged.fastq')\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read ID                                        1282289\n",
      "Sequence1                                      1282289\n",
      "Sequence2                                      1282289\n",
      "translation1                                   1282289\n",
      "translation2                                   1282289\n",
      "Average phredQ1                                1282289\n",
      "Average phredQ2                                1282289\n",
      "Number of bases with phredQ under 20 read 1    1282289\n",
      "Number of bases with phredQ under 20 read 2    1282289\n",
      "Is there a barcode in read 1?                  1282289\n",
      "Is there a barcode in read 2?                  1282289\n",
      "Nonbarcode read start position1                1282289\n",
      "Nonbarcode read start position2                1282289\n",
      "Barcode sequence1                              1282289\n",
      "Barcode sequence2                              1282289\n",
      "Mutation1                                      1282289\n",
      "Mutation2                                      1282289\n",
      "Number of mutations in read 1                  1282289\n",
      "Number of mutations in read 2                  1282289\n",
      "Barcode                                        1282289\n",
      "dtype: int64\n",
      "Read ID                                        979877\n",
      "Sequence1                                      979877\n",
      "Sequence2                                      979877\n",
      "translation1                                   979877\n",
      "translation2                                   979877\n",
      "Average phredQ1                                979877\n",
      "Average phredQ2                                979877\n",
      "Number of bases with phredQ under 20 read 1    979877\n",
      "Number of bases with phredQ under 20 read 2    979877\n",
      "Is there a barcode in read 1?                  979877\n",
      "Is there a barcode in read 2?                  979877\n",
      "Nonbarcode read start position1                979877\n",
      "Nonbarcode read start position2                979877\n",
      "Barcode sequence1                              979877\n",
      "Barcode sequence2                              979877\n",
      "Mutation1                                      979877\n",
      "Mutation2                                      979877\n",
      "Number of mutations in read 1                  979877\n",
      "Number of mutations in read 2                  979877\n",
      "Barcode                                        979877\n",
      "dtype: int64\n",
      "Read ID                                        877118\n",
      "Sequence1                                      877118\n",
      "Sequence2                                      877118\n",
      "translation1                                   877118\n",
      "translation2                                   877118\n",
      "Average phredQ1                                877118\n",
      "Average phredQ2                                877118\n",
      "Number of bases with phredQ under 20 read 1    877118\n",
      "Number of bases with phredQ under 20 read 2    877118\n",
      "Is there a barcode in read 1?                  877118\n",
      "Is there a barcode in read 2?                  877118\n",
      "Nonbarcode read start position1                877118\n",
      "Nonbarcode read start position2                877118\n",
      "Barcode sequence1                              877118\n",
      "Barcode sequence2                              877118\n",
      "Mutation1                                      877118\n",
      "Mutation2                                      877118\n",
      "Number of mutations in read 1                  877118\n",
      "Number of mutations in read 2                  877118\n",
      "Barcode                                        877118\n",
      "dtype: int64\n",
      "Read ID                                        203471\n",
      "Sequence1                                      203471\n",
      "Sequence2                                      203471\n",
      "translation1                                   203471\n",
      "translation2                                   203471\n",
      "Average phredQ1                                203471\n",
      "Average phredQ2                                203471\n",
      "Number of bases with phredQ under 20 read 1    203471\n",
      "Number of bases with phredQ under 20 read 2    203471\n",
      "Is there a barcode in read 1?                  203471\n",
      "Is there a barcode in read 2?                  203471\n",
      "Nonbarcode read start position1                203471\n",
      "Nonbarcode read start position2                203471\n",
      "Barcode sequence1                              203471\n",
      "Barcode sequence2                              203471\n",
      "Mutation1                                      203471\n",
      "Mutation2                                      203471\n",
      "Number of mutations in read 1                  203471\n",
      "Number of mutations in read 2                  203471\n",
      "Barcode                                        203471\n",
      "Single mutation                                203471\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(str(dfRaw.count()))\n",
    "print(str(dfQuality.count()))\n",
    "print(str(dfWTorSingle.count()))\n",
    "print(str(dfMutants.count()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "659012\n"
     ]
    }
   ],
   "source": [
    "#Here I take the parsed sequence files and find barcodes within them and generate a text file with barcodes\n",
    "\n",
    "realDataParsed = SeqIO.parse(\"NP-11-19-6-merged.fastq\", \"fastq\")\n",
    "barcodeCounter = 0\n",
    "barcode = None\n",
    "with open('barcodes_test_NP_11_19_6.txt', 'w') as openFile:\n",
    "    for record in realDataParsed:\n",
    "        if 'CCTGCAGGTTATTA' in record.seq:\n",
    "            barcodeCounter += 1\n",
    "            barcodeStart = record.seq.find('CCTGCAGGTTATTA')\n",
    "            barcode = record.seq[barcodeStart - 15:barcodeStart]        \n",
    "            openFile.write(str(barcode + '\\n'))\n",
    "\n",
    "print(barcodeCounter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1230141\n",
      "4616\n",
      "8726\n",
      "15799\n",
      "26637\n",
      "40794\n",
      "56460\n",
      "71946\n",
      "89990\n",
      "105835\n",
      "114727\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "\n",
    "fh = open('barcodes_test_NP_11_19_5.txt')\n",
    "barcodeList = fh.readlines()\n",
    "print(len(barcodeList))\n",
    "\n",
    "barcode_counts = Counter(barcodeList)\n",
    "\n",
    "#Counter(openfile).keys() # equals to list(set(words))\n",
    "print(len(Counter(barcodeList[:5000]).keys()))\n",
    "print(len(Counter(barcodeList[:10000]).keys()))\n",
    "print(len(Counter(barcodeList[:20000]).keys()))\n",
    "print(len(Counter(barcodeList[:40000]).keys()))\n",
    "print(len(Counter(barcodeList[:80000]).keys()))\n",
    "print(len(Counter(barcodeList[:160000]).keys()))\n",
    "print(len(Counter(barcodeList[:320000]).keys()))\n",
    "print(len(Counter(barcodeList[:640000]).keys()))\n",
    "print(len(Counter(barcodeList[:1000000]).keys()))\n",
    "print(len(Counter(barcodeList).keys()))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this bit should let us find every codon at a given position\n",
    "fh = open('NP_11_18_13.txt')\n",
    "openfile = fh.read()\n",
    "\n",
    "codonOILeftFlank = 'CTTGACTAC'\n",
    "\n",
    "numberofcodons = openfile.count(codonOILeftFlank)\n",
    "#barcodeLocation = openfile.find(barcodeRightFlank, 100000)\n",
    "print('There are ' + str(numberofcodons) + ' codons and they are:')\n",
    "#print(barcodeLocation)\n",
    "#print(openfile[barcodeLocation-20:barcodeLocation+10])\n",
    "newstart = 0\n",
    "for position in range(int(len(openfile))):\n",
    "    position = openfile.find(codonOILeftFlank, newstart)\n",
    "    #print(position)\n",
    "#    if '\\n' in openfile[position:position+4]:\n",
    "#        newstart = position + 1\n",
    "#        position = position + 1\n",
    "    if openfile[position+1:position+4] != 'GCC':\n",
    "        print(openfile[position+1:position+4])\n",
    "        newstart = position + 1\n",
    "        position = position + 1\n",
    "#    else:\n",
    "#        print(openfile[position+1:position+4])\n",
    "#        newstart = position + 1\n",
    "#        position = position + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-05117ae0e473>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;31m#f.write('the duplicate barcodes are')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbarcodeLines\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mbarcodeLines\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m         \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mbarcodeLines\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mline\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#first I need to make a list of barcodes that appear more than once\n",
    "\n",
    "fh = open('NP_11_18_13_barcodes.txt')\n",
    "barcodeLines = fh.readlines()\n",
    "f= open(\"NP_11_16_3_duplicateBarcodes.txt\",\"w+\")\n",
    "#f.write('the duplicate barcodes are')\n",
    "for line,i in enumerate(barcodeLines):\n",
    "    if i in barcodeLines[0:line]:\n",
    "        f.write(i)\n",
    "    if i in barcodeLines[line + 1:]:\n",
    "        f.write(i)\n",
    "print('done')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "#I want to make a new file that only contains reads with barcodes that appear\n",
    "#more than once along with their paired read\n",
    "\n",
    "barcodeLeftFlank = 'TTAGGAATTAATCATGG'\n",
    "barcode = None\n",
    "\n",
    "fh = open('NP_11_16_3.txt')\n",
    "fastqLines = fh.readlines()\n",
    "fh.close()\n",
    "\n",
    "fh = open('NP_11_16_3_duplicateBarcodes.txt')\n",
    "dupicateBarcodes = fh.read()\n",
    "fh.close()\n",
    "\n",
    "fh = open('NP_11_16_3_fastqDuplicatesOnlytest.txt', 'w+')\n",
    "\n",
    "for lineNumber,line in enumerate(fastqLines[:50000]):\n",
    "    if barcodeLeftFlank in line:\n",
    "        position = line.find(barcodeLeftFlank)\n",
    "        barcode = line[position + 17:position + 32]\n",
    "        if barcode != 'ATCCTGCAGGCTATT':\n",
    "            if barcode in dupicateBarcodes:\n",
    "                readNumberPosition = fastqLines[lineNumber-1].find(':N:') - 1\n",
    "                fh.writelines('\\nBarcode ' + barcode + ' has the following paired reads\\n')\n",
    "                if fastqLines[lineNumber-1][readNumberPosition] == '2':\n",
    "                    fh.writelines(fastqLines[lineNumber-5:lineNumber-1])\n",
    "                fh.writelines(fastqLines[lineNumber-1:lineNumber+3])\n",
    "                if fastqLines[lineNumber-1][readNumberPosition] == '1':\n",
    "                    fh.writelines(fastqLines[lineNumber+3:lineNumber+7])\n",
    "\n",
    "                instance = 0\n",
    "                for lineNumber2,line2 in enumerate(fastqLines):\n",
    "                    if barcode in line2:\n",
    "                        if lineNumber != lineNumber2:\n",
    "                            readNumberPosition2 = fastqLines[lineNumber2-1].find(':N:') - 1\n",
    "                            position2 = line2.find(barcodeLeftFlank)\n",
    "                            barcodeRepeat = line2[position2 + 17:position2 + 32]\n",
    "                            fh.writelines('\\nThe next instance of barcode ' + barcodeRepeat + ' has the following paired reads \\n')\n",
    "                            if fastqLines[lineNumber2-1][readNumberPosition2] == '2':\n",
    "                                fh.writelines(fastqLines[lineNumber2-5:lineNumber2-1])\n",
    "                            fh.writelines(fastqLines[lineNumber2-1:lineNumber2+3])\n",
    "                            if fastqLines[lineNumber2-1][readNumberPosition2] == '1':\n",
    "                                fh.writelines(fastqLines[lineNumber2+3:lineNumber2+7])\n",
    "                        instance = instance + 1\n",
    "                fh.writelines('\\nThere were ' + str(instance) + ' instances of this barcode \\n')\n",
    "\n",
    "print('done') \n",
    "fh.close()\n",
    "\n",
    "\n",
    "    \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
