{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For PacBio I need to set uop the analysis rather differently\n",
    "I used the following tutorial to do alignments and filtering\n",
    "https://jbloomlab.github.io/alignparse/lasv_pilot.html\n",
    "Unfortunately it doesn't work with no filtering so I just tried to set the filtering very loose\n",
    "That resulted in ~10% of reads being filtered, they are likely the \"worst\" 10%\n",
    "Jack suggested using consensus to figure out what mutations are real\n",
    " For barcodes with >2 reads we can just check if associated mutations appear in most of the barcode reads\n",
    " Only accept mutations that appear in a majority (>90%?)\n",
    "Then we can filter on there being just one, allowed mutation\n",
    "At that point we can basically move on to statistics and use the same ones generated in NP-11-19 and 24\n",
    "\n",
    "Add backbone stuff in\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import tempfile\n",
    "import warnings\n",
    "\n",
    "import Bio.SeqIO\n",
    "\n",
    "import pandas as pd\n",
    "import numpy\n",
    "\n",
    "from plotnine import *\n",
    "\n",
    "import pysam\n",
    "\n",
    "import alignparse.ccs\n",
    "import alignparse.minimap2\n",
    "import alignparse.targets\n",
    "import alignparse.cs_tag\n",
    "import alignparse.consensus\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NP_11_21_1              5719 bp ds-DNA     circular     07-MAY-2021\n",
      "DEFINITION  .\n",
      "KEYWORDS    \"Personalized Plasmid ID\" \"Plasmid Name\" \"Description\" \"Selection\" \n",
      "            \"Concentration (ng/uL)\" \"Location\" \"Sequenced verified?\" \"Date\n",
      "            Added\" \"Glycerol Stock?\"\n",
      "FEATURES             Location/Qualifiers\n",
      "     gene            1..1404\n",
      "                     /label=\"rbcLCodonOpt\"\n",
      "     CDS             1..1404\n",
      "                     /label=\"Translation 1-1404\"\n",
      "     Barcode         1426..1440\n",
      "                     /label=\"Barcode\"\n",
      "     backbone        1441..5719\n",
      "                     /label=\"backbone\"\n",
      "ORIGIN\n",
      "        1 ATGGACCAGA GTTCTCGCTA TGTCAATCTT GCTTTGAAAG AAGAGGACTT AATTGCCGGA\n",
      "       61 GGCGAACACG TATTGTGCGC CTATATTATG AAACCAAAAG CTGGGTACGG TTACGTTGCC\n",
      "      121 ACCGCAGCGC ACTTTGCTGC AGAGAGTTCG ACTGGAACAA ATGTAGAGGT GTGTACCACT\n",
      "      181 GATGATTTTA CTCGTGGCGT GGATGCCTTA GTTTACGAAG TAGATGAGGC CCGCGAGCTT\n",
      "      241 ACTAAGATTG CCTACCCAGT AGCACTGTTC GATCGTAATA TCACGGATGG AAAAGCTATG\n",
      "      301 ATCGCATCAT TTCTTACATT GACCATGGGC AATAATCAAG GCATGGGTGA CGTAGAATAT\n",
      "      361 GCCAAAATGC ATGACTTCTA TGTGCCTGAA GCGTATCGCG CTCTGTTTGA TGGCCCTTCT\n",
      "      421 GTAAATATTT CCGCCTTGTG GAAGGTACTG GGTCGTCCGG AGGTTGATGG AGGCTTAGTT\n",
      "      481 GTCGGTACTA TCATTAAACC AAAGCTTGGG CTGCGCCCTA AGCCTTTCGC TGAGGCTTGC\n",
      "      541 CACGCATTTT GGTTAGGTGG TGATTTTATT AAAAATGATG AACCGCAAGG GAACCAACCC\n",
      "      601 TTCGCGCCAC TTCGCGACAC GATCGCCTTG GTTGCAGATG CGATGCGCCG TGCACAGGAC\n",
      "      661 GAAACGGGTG AAGCAAAGTT GTTTAGTGCT AATATCACCG CTGATGACCC GTTTGAAATC\n",
      "      721 ATCGCCCGCG GGGAATACGT TCTTGAAACA TTCGGTGAAA ATGCAAGTCA CGTTGCCTTA\n",
      "      781 TTAGTAGACG GATACGTAGC CGGCGCAGCT GCTATTACGA CCGCTCGTCG CCGTTTCCCC\n",
      "      841 GATAATTTCT TGCATTATCA TCGCGCCGGG CATGGAGCGG TCACTAGCCC TCAGTCTAAG\n",
      "      901 CGCGGCTATA CTGCTTTTGT GCATTGTAAG ATGGCTCGTT TGCAAGGTGC GTCGGGTATC\n",
      "      961 CATACCGGAA CTATGGGCTT TGGAAAGATG GAAGGGGAGT CCTCAGATCG TGCCATCGCG\n",
      "     1021 TATATGTTAA CTCAAGATGA AGCACAAGGC CCCTTTTATC GCCAGTCGTG GGGAGGGATG\n",
      "     1081 AAGGCCTGCA CCCCTATTAT TTCTGGCGGA ATGAATGCCT TACGCATGCC CGGCTTCTTC\n",
      "     1141 GAAAATCTTG GTAATGCGAA TGTCATCCTG ACTGCTGGTG GCGGTGCTTT CGGCCATATT\n",
      "     1201 GACGGCCCTG TGGCCGGCGC TCGTTCTCTT CGCCAAGCTT GGCAGGCGTG GCGCGATGGC\n",
      "     1261 GTCCCTGTGC TTGACTACGC CCGCGAGCAT AAGGAGTTGG CACGTGCTTT TGAGTCCTTC\n",
      "     1321 CCTGGCGACG CTGACCAGAT TTATCCAGGT TGGCGTAAGG CGCTTGGAGT GGAGGATACT\n",
      "     1381 CGTTCAGCTT TACCTGCGTA ATAGCCTGCA GGACCGGTGG TGACCNNNNN NNNNNNNNNN\n",
      "     1441 CCATGCTTAA GCAATTGATT AATTCCTAAT TTTTGTTGAC ACTCTATCAT TGATAGAGTT\n",
      "     1501 ATTTTACCAC TCCCTATCAG TGATAGAGAA AAGTGAAAAA AAGCCAGCCA GTAGTAAGAG\n",
      "     1561 GGGGAATTCC GCatgcatca tcaccatcac cacagcaagc cagatcgtgt tgttttgatc\n",
      "     1621 ggcgttgccg gtgactccgg ttgcggcaaa tcaaccttcc taaatcgcct tgccgacttg\n",
      "     1681 tttggtacgg aattgatgac ggtcatctgc ttggatgact atcacagtct cgatcgcaag\n",
      "     1741 ggccggaagg aagcaggcgt aacggctttg gatccccgcg ccaacaactt tgacttgatg\n",
      "     1801 tatgaacagg tcaaggcgtt gaagaacggc gaaacgatca tgaagccgat ctacaaccat\n",
      "     1861 gaaaccggct tgatcgatcc gcccgaaaaa atcgaaccca atcgcatcat tgtgatcgag\n",
      "     1921 ggtctgcatc cgctttacga cgagcgcgtg cgtgaactgc tcgatttcag cgtttacctc\n",
      "     1981 gacatcgatg acgaagtcaa aatcgcttgg aagatccaac gcgatatggc agaacgcggc\n",
      "     2041 cactcctacg aagatgtcct cgcctcgatc gaagcgcgcc gccctgactt caaggcctac\n",
      "     2101 attgagcccc agcgtggcca tgcggacatc gtcatccgcg tcatgccgac ccagctaatc\n",
      "     2161 cccaatgaca ccgagcgcaa ggtgctgcgg gtgcagttga tccaacggga aggccgcgat\n",
      "     2221 ggttttgagc cggcttacct gttcgacgaa ggttcgacca tccagtggac gccctgcggt\n",
      "     2281 cgtaagctga cctgctccta tccgggcatt cgcttagcct acggccctga cacctactac\n",
      "     2341 ggtcacgaag tctcagtgct tgaggtcgac ggtcagttcg agaacctcga agagatgatc\n",
      "     2401 tacgtcgagg gccacctcag caagaccgac acgcagtact acggtgagtt gacccacctg\n",
      "     2461 ctgctacagc acaaagatta cccgggttcg aacaacggca cgggtctgtt ccaagtgctg\n",
      "     2521 accggcctga aaatgcgggc ggcctatgag cgtttgacct cccaagcagc acccGTCGCC\n",
      "     2581 GCTAGTGTCT AGATGgaatt cctgcagccc gggggatccc atggtacgcg tgctagaggc\n",
      "     2641 atcaaataaa acgaaaggct cagtcgaaag actgggcctt tcgttttatc tgttgtttgt\n",
      "     2701 cggtgaacgc tctcctgagt aggacaaatc cgccgccGCT AGCAGTTCGT AGACGGAAAC\n",
      "     2761 AAACGCAGAA TCCAAGCAAA AGGGAATAAG GGCGACACGG AAATGTTGAA TACTCATACT\n",
      "     2821 CTTCCTTTTT CAATATTATT GAAGCATTTA TCAGGGTTAT GTCCACTGGG TTCGTGCCTT\n",
      "     2881 CATCCGGATA TCCAGGAACT GTTAAGACCC ACTTTCACAT TTAAGTTGTT TTTCTAATCC\n",
      "     2941 GCATATGATC AATTCAAGGC CGAATAAGAA GGCTGGCTCT GCACCTTGGT GATCAAATAA\n",
      "     3001 TTCGATAGCT TGTCGTAATA ATGGCGGCAT ACTATCAGTA GTAGGTGTTT CCCTTTCTTC\n",
      "     3061 TTTAGCGACT TGATGCTCTT GATCTTCCAA TACGCAACCT AAAGTAAAAT GCCCCACAGC\n",
      "     3121 GCTGAGTGCA TATAATGCAT TCTCTAGTGA AAAACCTTGT TGGCATAAAA AGGCTAATTG\n",
      "     3181 ATTTTCGAGA GTTTCATACT GTTTTTCTGT AGGCCGTGTA CCTAAATGTA CTTTTGCTCC\n",
      "     3241 ATCGCGATGA CTTAGTAAAG CACATCTAAA ACTTTTAGCG TTATTACGTA AAAAATCTTG\n",
      "     3301 CCAGCTTTCC CCTTCTAAAG GGCAAAAGTG AGTATGGTGC CTATCTAACA TCTCAATGGC\n",
      "     3361 TAAGGCGTCG AGCAAAGCCC GCTTATTTTT TACATGCCAA TACAATGTAG GCTGCTCTAC\n",
      "     3421 ACCTAGCTTC TGGGCGAGTT TACGGGTTGT TAAACCTTCG ATTCCGACCT CATTAAGCAG\n",
      "     3481 CTCTAATGCG CTGTTAATCA CTTTACTTTT ATCTAATCTA GACATactct tcctttttca\n",
      "     3541 atattattga agcatttatc agggttattg tctcatgagc ggatacatat ttgaatgtat\n",
      "     3601 ttagaaaaat aaacaaatag gggttccgcg TTTCCCTCGA CAATTCGCGc tagacctagg\n",
      "     3661 gcgttcggct gcggcgagcg gtatcagctc actcaaaggc ggtaatacgg ttatccacag\n",
      "     3721 aatcagggga taacgcagga aagaacatgt gagcaaaagg ccagcaaaag gccaggaacc\n",
      "     3781 gtaaaaaggc cgcgttgctg gcgtttttcc ataggctccg cccccctgac gagcatcaca\n",
      "     3841 aaaatcgacg ctcaagtcag aggtggcgaa acccgacagg actataaaga taccaggcgt\n",
      "     3901 ttccccctgg aagctccctc gtgcgctctc ctgttccgac cctgccgctt accggatacc\n",
      "     3961 tgtccgcctt tctcccttcg ggaagcgtgg cgctttctca TAgctcacgc tgtaggtatc\n",
      "     4021 tcagttcggt gtaggtcgtt cgctccaagc tgggctgtgt gcacgaaccc cccgttcagc\n",
      "     4081 ccgaccgctg cgccttatcc ggtaactatc gtcttgagtc caacccggta agacacgact\n",
      "     4141 tatcgccact ggcagcagcc actggtaaca ggattagcag agcgaggtat gtaggcggtg\n",
      "     4201 ctacagagtt cttgaagtgg tggcctaact acggctacac tagaaggaca gtatttggta\n",
      "     4261 tctgcgctct gctgaagcca gttaccttcg gaaaaagagt tggtagctct tgatccggca\n",
      "     4321 aacaaaccac cgctggtagc ggtggttttt ttgtttgcaa gcagcagatt acgcgcagaa\n",
      "     4381 aaaaaggatc tcaagaagat cctttgatct tttctacggg gtctgacgct cagtggaacg\n",
      "     4441 aaaactcacg ttaagggatt ttggtcatga ctagtgcttg gattctcacc aataaaaaac\n",
      "     4501 gcccggcggc aaccgagcgt tctgaacaaa tccagatgga gttctgaggt cattactgga\n",
      "     4561 tctatcaaca ggagtccaag cgagctctcg aaccccagag tcccgctcag aagaactcgt\n",
      "     4621 caagaaggcg atagaaggcg atgcgctgcg aatcgggagc ggcgataccg taaagcacga\n",
      "     4681 ggaagcggtc agcccattcg ccgccaagct cttcagcaat atcacgggta gccaacgcta\n",
      "     4741 tgtcctgata gcggtccgcc acacccagcc ggccacagtc gatgaatcca gaaaagcggc\n",
      "     4801 cattttccac catgatattc ggcaagcagg catcgccatg ggtcacgacg agatcctcgc\n",
      "     4861 cgtcgggcat gcgcgccttg agcctggcga acagttcggc tggcgcgagc ccctgatgct\n",
      "     4921 cttcgtccag atcatcctga tcgacaagac cggcttccat ccgagtacgt gctcgctcga\n",
      "     4981 tgcgatgttt cgcttggtgg tcgaatgggc aggtagccgg atcaagcgta tgcagccgcc\n",
      "     5041 gcattgcatc agccatgatg gatactttct cggcaggagc aaggtgagat gacaggagat\n",
      "     5101 cctgccccgg cacttcgccc aatagcagcc agtcccttcc cgcttcagtg acaacgtcga\n",
      "     5161 gcacagctgc gcaaggaacg cccgtcgtgg ccagccacga tagccgcgct gcctcgtcct\n",
      "     5221 gcagttcatt cagggcaccg gacaggtcgg tcttgacaaa aagaaccggg cgcccctgcg\n",
      "     5281 ctgacagccg gaacacggcg gcatcagagc agccgattgt ctgttgtgcc cagtcatagc\n",
      "     5341 cgaatagcct ctccacccaa gcggccggag aacctgcgtg caatccatct tgttcaatca\n",
      "     5401 tgcgaaacga tcctcatcct gtctcttgat cagatcttga tcccctgcgc catcagatcc\n",
      "     5461 ttggcggcaa gaaagccatc cagtttactt tgcagggctt cccaacctta ccagagggcg\n",
      "     5521 ccccagctgg caattccgac gtctaagaaa ccattattat catgacatta acctataaaa\n",
      "     5581 ataggcgtat cacgaggccc tttcgtcttc acctcgagtc cctatcagtg atagagattg\n",
      "     5641 acatccctat cagtgataga gatactgagc acatcagcag gacgcactga ccgaattcat\n",
      "     5701 taaagaggag aaaggtacc\n",
      "//\n"
     ]
    }
   ],
   "source": [
    "# First I make a template based on the plasmid with only the features we care about\n",
    "\n",
    "with open('np_11_21_1_bb.gb') as f:\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_11_21_1:\n",
      "  query_clip5: null\n",
      "  query_clip3: null\n",
      "  backbone:\n",
      "    return:\n",
      "    - mutations\n",
      "    - accuracy\n",
      "    filter:\n",
      "      mutation_nt_count: null\n",
      "      mutation_op_count: null\n",
      "      clip5: null\n",
      "      clip3: null\n",
      "  gene:\n",
      "    filter:\n",
      "      mutation_nt_count: null\n",
      "      mutation_op_count: null\n",
      "      clip5: null\n",
      "      clip3: null\n",
      "    return:\n",
      "    - mutations\n",
      "    - accuracy\n",
      "  Barcode:\n",
      "    return:\n",
      "    - sequence\n",
      "    filter:\n",
      "      clip5: 0\n",
      "      clip3: 0\n",
      "      mutation_nt_count: 0\n",
      "      mutation_op_count: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# The Bloom lab uses something called a YAML file that tells the aligner how to parse features\n",
    "# I edited the one they use for Lassa to fit rbcL, this will return mutations in rbcL and the sequence of the barcode\n",
    "\n",
    "parse_specs_file = 'rbcLDMSbb.yaml'\n",
    "    \n",
    "targets = alignparse.targets.Targets(seqsfile = 'np_11_21_1_bb.gb', \\\n",
    "                                     feature_parse_specs = parse_specs_file, \\\n",
    "                                     allow_extra_features = True, \\\n",
    "                                     allow_clipped_muts_seqs = True)\n",
    "\n",
    "print(targets.feature_parse_specs('yaml'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This chooses minimap2 as the aligner\n",
    "\n",
    "mapper = alignparse.minimap2.Mapper(alignparse.minimap2.OPTIONS_CODON_DMS, \\\n",
    "                                    prog='minimap2-2.17-hbbe82c9_3/bin/minimap2')\n",
    "\n",
    "# It's possible to do multiple targets and have different categories of files \n",
    "# This just does one queryfile (the fastq file with the data) and one template ('targets' defined above)\n",
    "# The output is a SAM file\n",
    "\n",
    "targets.align(queryfile=\"/Volumes/NP_DFS_4TB/NP_11_25/subreads_ccs.fastq\", \\\n",
    "              alignmentfile='rbcLalignmentsBB.sam', \\\n",
    "              mapper=mapper)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aligned NP_11_21_1</td>\n",
       "      <td>3585880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>filtered NP_11_21_1</td>\n",
       "      <td>184045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>unmapped</td>\n",
       "      <td>79734</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              category    count\n",
       "0   aligned NP_11_21_1  3585880\n",
       "1  filtered NP_11_21_1   184045\n",
       "2             unmapped    79734"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This figures out what aligned and what was filtered out\n",
    "\n",
    "run_readstats, run_aligned, run_filtered = targets.parse_alignment('rbcLalignmentsBB.sam', filtered_cs=False)\n",
    "run_readstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_name</th>\n",
       "      <th>filter_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m64044_201203_021636/1/ccs</td>\n",
       "      <td>Barcode clip5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m64044_201203_021636/14/ccs</td>\n",
       "      <td>Barcode mutation_nt_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m64044_201203_021636/47/ccs</td>\n",
       "      <td>Barcode clip5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m64044_201203_021636/72/ccs</td>\n",
       "      <td>Barcode clip5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m64044_201203_021636/101/ccs</td>\n",
       "      <td>Barcode mutation_nt_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184040</th>\n",
       "      <td>m64044_201203_021636/180554275/ccs</td>\n",
       "      <td>Barcode clip5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184041</th>\n",
       "      <td>m64044_201203_021636/180554276/ccs</td>\n",
       "      <td>Barcode mutation_nt_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184042</th>\n",
       "      <td>m64044_201203_021636/180554327/ccs</td>\n",
       "      <td>Barcode clip5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184043</th>\n",
       "      <td>m64044_201203_021636/180554512/ccs</td>\n",
       "      <td>Barcode clip5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184044</th>\n",
       "      <td>m64044_201203_021636/180554542/ccs</td>\n",
       "      <td>Barcode mutation_nt_count</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>184045 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                query_name              filter_reason\n",
       "0               m64044_201203_021636/1/ccs              Barcode clip5\n",
       "1              m64044_201203_021636/14/ccs  Barcode mutation_nt_count\n",
       "2              m64044_201203_021636/47/ccs              Barcode clip5\n",
       "3              m64044_201203_021636/72/ccs              Barcode clip5\n",
       "4             m64044_201203_021636/101/ccs  Barcode mutation_nt_count\n",
       "...                                    ...                        ...\n",
       "184040  m64044_201203_021636/180554275/ccs              Barcode clip5\n",
       "184041  m64044_201203_021636/180554276/ccs  Barcode mutation_nt_count\n",
       "184042  m64044_201203_021636/180554327/ccs              Barcode clip5\n",
       "184043  m64044_201203_021636/180554512/ccs              Barcode clip5\n",
       "184044  m64044_201203_021636/180554542/ccs  Barcode mutation_nt_count\n",
       "\n",
       "[184045 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# here we find out why stuff was filtered\n",
    "\n",
    "run_filtered['NP_11_21_1']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_name</th>\n",
       "      <th>query_clip5</th>\n",
       "      <th>query_clip3</th>\n",
       "      <th>backbone_mutations</th>\n",
       "      <th>backbone_accuracy</th>\n",
       "      <th>gene_mutations</th>\n",
       "      <th>gene_accuracy</th>\n",
       "      <th>Barcode_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m64044_201203_021636/2/ccs</td>\n",
       "      <td>2404</td>\n",
       "      <td>0</td>\n",
       "      <td>del173to173 T1874C C1875T</td>\n",
       "      <td>0.999810</td>\n",
       "      <td>ins837C C1130G C1131T</td>\n",
       "      <td>0.999530</td>\n",
       "      <td>TAAATGTTAAGGCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m64044_201203_021636/4/ccs</td>\n",
       "      <td>2384</td>\n",
       "      <td>0</td>\n",
       "      <td>del62to62 ins88G del127to127 ins207C ins327T d...</td>\n",
       "      <td>0.988948</td>\n",
       "      <td>del15to15 G36T A37G C57G G58C ins96A del212to2...</td>\n",
       "      <td>0.984765</td>\n",
       "      <td>AACACTTAACATCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m64044_201203_021636/5/ccs</td>\n",
       "      <td>2410</td>\n",
       "      <td>0</td>\n",
       "      <td>ins52G ins72C A281C ins372C ins392A del406to40...</td>\n",
       "      <td>0.993817</td>\n",
       "      <td>ins103G ins302T ins385C ins523C del730to730 A1...</td>\n",
       "      <td>0.993166</td>\n",
       "      <td>CTGCCCTCAATTAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m64044_201203_021636/6/ccs</td>\n",
       "      <td>2418</td>\n",
       "      <td>0</td>\n",
       "      <td>ins178A ins318C ins334C ins345C ins434T ins623...</td>\n",
       "      <td>0.997527</td>\n",
       "      <td>ins96A ins119C ins198C ins448C ins837CC ins868...</td>\n",
       "      <td>0.997315</td>\n",
       "      <td>TTAATAAGAGAAATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m64044_201203_021636/7/ccs</td>\n",
       "      <td>2404</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0.999866</td>\n",
       "      <td>C523G C524A T525A G645A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>AGAATCTACAATCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3585875</th>\n",
       "      <td>m64044_201203_021636/180554585/ccs</td>\n",
       "      <td>2405</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0.999731</td>\n",
       "      <td>C101A T102A</td>\n",
       "      <td>0.999928</td>\n",
       "      <td>TCTGGGGTGATGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3585876</th>\n",
       "      <td>m64044_201203_021636/180554586/ccs</td>\n",
       "      <td>2404</td>\n",
       "      <td>0</td>\n",
       "      <td>del349to349</td>\n",
       "      <td>0.999894</td>\n",
       "      <td>C265T T266G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>TCTTTAGGTCCACGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3585877</th>\n",
       "      <td>m64044_201203_021636/180554587/ccs</td>\n",
       "      <td>2394</td>\n",
       "      <td>0</td>\n",
       "      <td>del120to120 del446to446 del480to480 del578to57...</td>\n",
       "      <td>0.997998</td>\n",
       "      <td>ins124G G295T C296G T297C del564to564 del666to...</td>\n",
       "      <td>0.997055</td>\n",
       "      <td>TATTCTCCTTCCTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3585878</th>\n",
       "      <td>m64044_201203_021636/180554589/ccs</td>\n",
       "      <td>2394</td>\n",
       "      <td>0</td>\n",
       "      <td>del96to96 C198G G200T del222to222 ins761G del1...</td>\n",
       "      <td>0.997614</td>\n",
       "      <td>del80to80 del255to255 del291to291 del327to327 ...</td>\n",
       "      <td>0.996862</td>\n",
       "      <td>GCTTGCTGAATAGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3585879</th>\n",
       "      <td>m64044_201203_021636/180554591/ccs</td>\n",
       "      <td>2404</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1.000000</td>\n",
       "      <td>del529to569 G616T A617G C618G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>GAGGGTGGTCCGGGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3585880 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 query_name  query_clip5  query_clip3  \\\n",
       "0                m64044_201203_021636/2/ccs         2404            0   \n",
       "1                m64044_201203_021636/4/ccs         2384            0   \n",
       "2                m64044_201203_021636/5/ccs         2410            0   \n",
       "3                m64044_201203_021636/6/ccs         2418            0   \n",
       "4                m64044_201203_021636/7/ccs         2404            0   \n",
       "...                                     ...          ...          ...   \n",
       "3585875  m64044_201203_021636/180554585/ccs         2405            0   \n",
       "3585876  m64044_201203_021636/180554586/ccs         2404            0   \n",
       "3585877  m64044_201203_021636/180554587/ccs         2394            0   \n",
       "3585878  m64044_201203_021636/180554589/ccs         2394            0   \n",
       "3585879  m64044_201203_021636/180554591/ccs         2404            0   \n",
       "\n",
       "                                        backbone_mutations  backbone_accuracy  \\\n",
       "0                                del173to173 T1874C C1875T           0.999810   \n",
       "1        del62to62 ins88G del127to127 ins207C ins327T d...           0.988948   \n",
       "2        ins52G ins72C A281C ins372C ins392A del406to40...           0.993817   \n",
       "3        ins178A ins318C ins334C ins345C ins434T ins623...           0.997527   \n",
       "4                                                                    0.999866   \n",
       "...                                                    ...                ...   \n",
       "3585875                                                              0.999731   \n",
       "3585876                                        del349to349           0.999894   \n",
       "3585877  del120to120 del446to446 del480to480 del578to57...           0.997998   \n",
       "3585878  del96to96 C198G G200T del222to222 ins761G del1...           0.997614   \n",
       "3585879                                                              1.000000   \n",
       "\n",
       "                                            gene_mutations  gene_accuracy  \\\n",
       "0                                    ins837C C1130G C1131T       0.999530   \n",
       "1        del15to15 G36T A37G C57G G58C ins96A del212to2...       0.984765   \n",
       "2        ins103G ins302T ins385C ins523C del730to730 A1...       0.993166   \n",
       "3        ins96A ins119C ins198C ins448C ins837CC ins868...       0.997315   \n",
       "4                                  C523G C524A T525A G645A       1.000000   \n",
       "...                                                    ...            ...   \n",
       "3585875                                        C101A T102A       0.999928   \n",
       "3585876                                        C265T T266G       1.000000   \n",
       "3585877  ins124G G295T C296G T297C del564to564 del666to...       0.997055   \n",
       "3585878  del80to80 del255to255 del291to291 del327to327 ...       0.996862   \n",
       "3585879                      del529to569 G616T A617G C618G       1.000000   \n",
       "\n",
       "        Barcode_sequence  \n",
       "0        TAAATGTTAAGGCGT  \n",
       "1        AACACTTAACATCTG  \n",
       "2        CTGCCCTCAATTAGA  \n",
       "3        TTAATAAGAGAAATG  \n",
       "4        AGAATCTACAATCCC  \n",
       "...                  ...  \n",
       "3585875  TCTGGGGTGATGGGG  \n",
       "3585876  TCTTTAGGTCCACGT  \n",
       "3585877  TATTCTCCTTCCTTT  \n",
       "3585878  GCTTGCTGAATAGCC  \n",
       "3585879  GAGGGTGGTCCGGGC  \n",
       "\n",
       "[3585880 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here we can see mutations and barcodes and make a csv file for future use\n",
    "\n",
    "\n",
    "\n",
    "# Change filename to have automated experiment name added\n",
    "run_aligned['NP_11_21_1'].to_csv('NP_11_21_1_backboneAware.csv')\n",
    "\n",
    "run_aligned['NP_11_21_1']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
