{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PacBio analysis pipeline\n",
    "I used the following tutorial to do alignments and filtering: https://jbloomlab.github.io/alignparse/lasv_pilot.html\n",
    "\n",
    "Pipeline:\n",
    "- Filter with dmstools2, lose <10% (mostly for barcodes that don't align)\n",
    "- Make two dataframes, one with a row for each read and one with a row for each unique barcode\n",
    "- Make a consensus mutation for each barcode from all reads of that barcode\n",
    "    -  This depends on there being enough reads of an individual barcode\n",
    "    -  If a mutation appears in <X% of reads of a given barcode chuck it\n",
    "- See if what remains is an allowed mutation, if so keep it\n",
    "- Look for mutations in the backbone, filter barcodes with such mutations\n",
    "- Only keep barcodes with >Y reads\n",
    "\n",
    "### Defined Functions\n",
    "**mutConsenses:**\n",
    "    - Makes a list of all mutants and a list of just unique mutations\n",
    "    - figures out what fraction of reads each mutation is in\n",
    "    - Makes final single member list for easy conversion\n",
    "**enumerateMuts:**\n",
    "    - This takes the unique barcodes and makes a list of all of the barcodes that are one off from it\n",
    "**oneCodonOnlyCheck:**\n",
    "    - Determine whether consensus mutation is in the same column\n",
    "**Translate:**\n",
    "    - Used to make the aminoacid notation mutation name\n",
    "**passedAllFilters:**\n",
    "    - Generate final barcode list with paired mutants after all filters\n",
    "**filterReads:**\n",
    "    - Filters we use to generate independent datasets\n",
    "**getBCfreqs:**\n",
    "    - Filters out no mutation reads\n",
    "    - Gets frequency of each barcode\n",
    "**getMutPos:**\n",
    "    - Extract all mutated nucleotide positions in a given consensus\n",
    "**getCodonSeqs:**\n",
    "    - Figure out what the original and the new codons are\n",
    "**OBOcheck:**\n",
    "    - Figure out which barcodes are one mutation away from another barcode\n",
    "    - Return the similar barcode.  Keep the barcode with more reads\n",
    "**getAAmut:**\n",
    "    - Return the mutation and the new aminoacid for barcodes with just one mutated codon\n",
    "**mutFilters:**\n",
    "    - Make columns indicating if there is a consensus mutation with no indels\n",
    "**BBmutFilters:**\n",
    "    - Choose barcodes where the backbone has either no mutation or no consensus mutation\n",
    "\n",
    "**Preassigned Variable List:** <br>\n",
    "    - filterLabels <br>\n",
    "    - PROGRAMMED_MUTATION_CODONS <br>\n",
    "    - Template sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T20:01:44.736821Z",
     "start_time": "2021-05-19T20:01:44.733290Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.8.8 (default, Apr 13 2021, 12:59:45) \n",
      "[Clang 10.0.0 ]\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(sys.version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T20:02:58.656466Z",
     "start_time": "2021-05-19T20:02:54.011055Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Import packages, set global parameters\n",
    "import numpy as np\n",
    "import itertools\n",
    "import math\n",
    "import pandas as pd\n",
    "import time\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from ast import literal_eval\n",
    "import csv\n",
    "\n",
    "# This is the degree of consensus we will be looking for later\n",
    "CONSENSUS_CUTOFF = 0.9\n",
    "\n",
    "# This is the minimum number of reads we will accept for the final lookup table\n",
    "MINIMUM_READS = 4\n",
    "\n",
    "# Total number of mutations we are looking for\n",
    "TOTAL_MUTS = 8778\n",
    "\n",
    "FILE_NAME = 'NP_11_21_1_BB_AB'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T20:03:00.526059Z",
     "start_time": "2021-05-19T20:03:00.501859Z"
    },
    "code_folding": [
     0,
     3,
     27,
     35,
     46,
     72,
     87,
     108,
     136,
     145,
     160,
     176,
     191,
     201,
     207
    ]
   },
   "outputs": [],
   "source": [
    "# Function definitions \n",
    "\n",
    "#---------------------------------------------------\n",
    "def mutConsenses(BCmutListList):\n",
    "    readsOfBC = len(BCmutListList)\n",
    "    consensusMutDNA = []\n",
    "    compiledMutList = []\n",
    "    consensusPercentBCs = []\n",
    "    compiledMutListUnique = []\n",
    "    \n",
    "    for mutList in BCmutListList:\n",
    "        if type(mutList) != list:\n",
    "            mutList = ['noMutation']\n",
    "        for mut in mutList:\n",
    "            compiledMutList.append(mut)\n",
    "            if mut not in compiledMutListUnique:\n",
    "                compiledMutListUnique.append(mut)\n",
    "    for uniqueMut in compiledMutListUnique:\n",
    "        percentBCs = compiledMutList.count(uniqueMut)/readsOfBC\n",
    "        if percentBCs > CONSENSUS_CUTOFF:\n",
    "            consensusMutDNA.append(uniqueMut)\n",
    "            consensusPercentBCs.append(percentBCs)\n",
    "    if consensusMutDNA == []:\n",
    "        consensusMutDNA = ['NoConsensus']\n",
    "        consensusPercentBCs = ['NoConsensus']\n",
    "    return consensusMutDNA, consensusPercentBCs\n",
    "#---------------------------------------------------\n",
    "def enumerateMuts(barcode):\n",
    "    offByOnes = []\n",
    "    for i, letter in enumerate(barcode):\n",
    "        for base in ['A','T','G','C']:\n",
    "            if letter != base:\n",
    "                offByOnes.append(barcode[:i] + base + barcode[i + 1:])\n",
    "    return offByOnes\n",
    "#----------------------------------------------------\n",
    "def oneCodonOnlyCheck(listOfPositions):\n",
    "    affectedCodons = []\n",
    "    for position in listOfPositions:\n",
    "        # position-1 gets the position with a 0 index\n",
    "        # dividing by 3 and taking the floor gets the codon position\n",
    "        # The codon position already has a 0 index\n",
    "        mutPosition = math.floor((position-1)/3)\n",
    "        if mutPosition not in (affectedCodons):\n",
    "            affectedCodons.append(math.floor((position-1)/3))\n",
    "    return(affectedCodons)\n",
    "#-----------------------------------------------------\n",
    "def translate(seq): \n",
    "    table = { \n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_', \n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W', \n",
    "    } \n",
    "    protein =\"\" \n",
    "    if len(seq)%3 == 0: \n",
    "        for i in range(0, len(seq), 3): \n",
    "            codon = seq[i:i + 3] \n",
    "            protein += table[codon] \n",
    "    return protein \n",
    "#-------------------------------------------------------\n",
    "def passedAllFilters (row):\n",
    "    if row['OBO_found'] == True or \\\n",
    "        row['No_Mutations'] == True or \\\n",
    "        row['No_Consensus'] == True or \\\n",
    "        row['Indels'] == True or \\\n",
    "        row['HowManyAffectedCodons'] != 1 or \\\n",
    "        row['Codon_Allowed'] == False or \\\n",
    "        row['mutantAA'] == 'Silent mutation' or \\\n",
    "        row['mutantAA'] == 'Nonsense mutation' or \\\n",
    "        row['Enough_reads'] == False or \\\n",
    "        row['No_BB_Mutations'] == False:\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "#----------------------------------------------------\n",
    "def filterReads(df):\n",
    "    filteredDFs = {}\n",
    "    noFilter = df['BCfrequency'] > 0\n",
    "    OBOCheck = df['OBO_found'] == False\n",
    "    noMutCheck = df['No_Mutations'] == False\n",
    "    consensusCheck = df['No_Consensus'] == False\n",
    "    indelCheck = df['Indels'] == False\n",
    "    OCOcheck = df['HowManyAffectedCodons'] == 1\n",
    "    allowedCheck = df['Codon_Allowed'] == True\n",
    "    silentCheck = df['AAMutation'] != 'Silent mutation'\n",
    "    nonsenseCheck = df['AAMutation'] != 'Nonsense mutation'\n",
    "    enoughReadsCheck = df['Enough_reads'] == True\n",
    "    BBMutCheck = df['No_BB_Mutations'] == True\n",
    "    allFilters = OBOCheck & noMutCheck & consensusCheck & indelCheck & OCOcheck & allowedCheck & \\\n",
    "                silentCheck & nonsenseCheck & enoughReadsCheck & BBMutCheck\n",
    "    filterSet = [noFilter, OBOCheck, noMutCheck, consensusCheck, indelCheck, OCOcheck, allowedCheck, \\\n",
    "                 silentCheck, nonsenseCheck, enoughReadsCheck, BBMutCheck, allFilters]\n",
    "    for i, aFilter in enumerate(filterSet):\n",
    "        filteredDFs[filterLabels[i]] = df[aFilter]\n",
    "    return filteredDFs\n",
    "#-----------------------------------------------------\n",
    "def getConsensus(uniqueBCsIn): # This also checks for consensus\n",
    "    consensusMutList = []\n",
    "    consensusLevelList = []\n",
    "    consensusBBMutList = []\n",
    "    appearanceList = []\n",
    "    for x, BC in enumerate(uniqueBCsIn):\n",
    "        consensusMut = 'XXX'\n",
    "        consensusLevel = 0\n",
    "        BCfrequency = 0\n",
    "        uniqueBCGroup = BCGroups.get_group(BC)\n",
    "        BCmutListList = uniqueBCGroup['splitMuts'].array\n",
    "        BCBBmutListList = uniqueBCGroup['splitBBMuts'].array\n",
    "#         appearanceList.append(uniqueBCGroup['BCfrequency'].array[0])\n",
    "        consensusMut, consensusLevel = mutConsenses(BCmutListList)\n",
    "        consensusMutList.append(consensusMut)\n",
    "        consensusLevelList.append(consensusLevel)\n",
    "        consensusBBMutList.append(mutConsenses(BCBBmutListList)[0])\n",
    "        if x in [0,500,1000,5000,10000,50000,100000,200000]:\n",
    "            print(str(100*x/len(uniqueBCsIn))[:4] + ' percent done')\n",
    "    consensusDict = {\n",
    "                      \"Barcode_sequence\": uniqueBCs,\n",
    "                      \"Consensus_mutations\": consensusMutList,\n",
    "#                       \"BCFrequency\": appearanceList,\n",
    "                      \"Backbone_consensus_muts\": consensusBBMutList\n",
    "                    }\n",
    "    newDF = pd.DataFrame.from_dict(consensusDict)\n",
    "    return newDF\n",
    "#-----------------------------------------------------\n",
    "def getMutPos(row):\n",
    "    positionList=[]\n",
    "    if (row['No_Mutations'] == False and row['No_Consensus'] == False and row['Indels'] == False):\n",
    "        for mut in row['Consensus_mutations']:\n",
    "            positionList.append(int(mut[1:-1]))\n",
    "    else:\n",
    "        positionList.append(-1)\n",
    "    return positionList\n",
    "#-----------------------------------------------------\n",
    "def getCodonSeqs (row):\n",
    "    originalCodon = 'XXX'\n",
    "    mutantCodon = 'YYY'\n",
    "    mutantCodonList = []\n",
    "    if row['HowManyAffectedCodons'] == 1 and row['Indels'] == False:\n",
    "        codonPosition = row['AffectedCodonInt']\n",
    "        DNAposition = codonPosition * 3\n",
    "        originalCodon = TEMPLATE_STRING[DNAposition:DNAposition + 3]\n",
    "        mutCodonSeqList = list(originalCodon)\n",
    "        for mutationNumber, naPosition in enumerate(row['Mut_positions']):\n",
    "            withinCodonPosition = (naPosition - 1)%3\n",
    "            mutCodonSeqList[withinCodonPosition] = row['Consensus_mutations'][mutationNumber][-1:]\n",
    "            mutantCodon = ''.join(mutCodonSeqList)\n",
    "    return originalCodon, mutantCodon\n",
    "#-----------------------------------------------------\n",
    "def OBOcheck(row): \n",
    "    uniqueBC = row['Barcode_sequence']\n",
    "    BCtoAdd = None\n",
    "    hammingDFound = False\n",
    "    for barcodeWithOneMistake in enumerateMuts(uniqueBC):\n",
    "        if barcodeWithOneMistake in uniqueBCLookup:\n",
    "            hammingDFound = True\n",
    "            BCtoAdd = barcodeWithOneMistake\n",
    "            # Since we are running from less to more frequent barcodes\n",
    "            # This removes the less frequent barcode from consideration\n",
    "            uniqueBCLookup.remove(uniqueBC)\n",
    "            break\n",
    "#     if row.name%50000 == 0:\n",
    "#         print(str(100*row.name/len(uniqueBCs))[:4] + ' percent done')\n",
    "    return hammingDFound, BCtoAdd\n",
    "#-----------------------------------------------------\n",
    "def getAAmut(row):\n",
    "    aaMut = 'X0Y'\n",
    "    mutantAA = 'X'\n",
    "    if (row['Original_codon'] != 'XXX') and (row['Mutant_codon'] != 'YYY'):\n",
    "        originalAA = translate(row['Original_codon'])\n",
    "        AAposition = row['AffectedCodonInt']\n",
    "        mutantAA = translate(row['Mutant_codon'])\n",
    "        if originalAA == mutantAA:\n",
    "            aaMut = mutantAA = 'Silent mutation'\n",
    "        elif mutantAA == '_':\n",
    "            aaMut = mutantAA = 'Nonsense mutation'\n",
    "        else:\n",
    "            aaMut = (originalAA + str(AAposition) + mutantAA)\n",
    "    return aaMut, mutantAA\n",
    "#-----------------------------------------------------\n",
    "def mutFilters(row):\n",
    "    noMut = noCon = indel = False\n",
    "    if row.Consensus_mutations == ['noMutation']:\n",
    "        noMut = True\n",
    "    if row.Consensus_mutations == ['NoConsensus']:\n",
    "        noCon = True\n",
    "    if ('ins' in str(row.Consensus_mutations)) or ('del' in str(row.Consensus_mutations)):\n",
    "        indel = True\n",
    "    return noMut, noCon, indel\n",
    "#-----------------------------------------------------\n",
    "def BBmutFilters(row):\n",
    "    noMut = False\n",
    "    if (row.Backbone_consensus_muts == ['noMutation']) or (row.Backbone_consensus_muts == ['NoConsensus']):\n",
    "        noMut = True\n",
    "    return noMut\n",
    "#-----------------------------------------------------\n",
    "def shiftAAone (row):\n",
    "    corrected = 'X0Y'\n",
    "    if row['AAMutation'][1:-1].isnumeric():\n",
    "        position = int(row['AAMutation'][1:-1])\n",
    "        corrected = row['AAMutation'][0] + str(position + 1) + row['AAMutation'][-1]\n",
    "    return corrected\n",
    "#============================================================================\n",
    "#============================================================================\n",
    "filterLabels = ['All data', 'Off-by-one barcodes', 'No mutation found', \\\n",
    "                    'No consensus found', 'Indel found', 'More than one codon mutated', \\\n",
    "                    'Unallowed mutation', 'Missense mutation', 'Nonsense mutation', \\\n",
    "                    'Not enough reads', 'Backbone Mutation', 'Fully filtered']\n",
    "#-----------------------------------------------------\n",
    "PROGRAMMED_MUTATION_CODONS = ['CGT','CAT','AAA','GAT','GAA',\n",
    "                              'AGC','ACC','AAT','CAG','TGC',\n",
    "                              'GGC','CCG','GCG','GTG','ATT',\n",
    "                              'CTG','ATG','TTT','TAT','TGG']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T20:03:33.797Z"
    },
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Import previously generated dataframe, add BC freq \n",
    "# This dataframe is PacBio data after using alignParse, it has barcodes, gene mutations and backbone mutations\n",
    "readsDF = pd.read_csv('NP_11_21_1_backboneAware.csv') \n",
    "\n",
    "# Add a new column that displays how many times each barcode appears in the list\n",
    "readsDF['BCfrequency'] = readsDF['Barcode_sequence'].map(readsDF['Barcode_sequence'].value_counts())\n",
    "readsDF[['query_name', 'gene_mutations', 'Barcode_sequence', 'BCfrequency']]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T20:03:35.421Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Convert values in mutation column into usable values \n",
    "# First split them by the spaces\n",
    "readsDF['splitMuts'] = readsDF['gene_mutations'].str.split()\n",
    "readsDF['splitBBMuts'] = readsDF['backbone_mutations'].str.split()\n",
    "\n",
    "# This does a groupby and also makes a list of unique barcodes\n",
    "BCGroups = readsDF.groupby('Barcode_sequence')\n",
    "uniqueBCs = readsDF.Barcode_sequence.unique()\n",
    "\n",
    "# uniqueBCsAndFreq = BCGroups.apply(lambda x: x['BCfrequency'].unique())\n",
    "print('There are ' + str(readsDF.Barcode_sequence.nunique()) + ' unique barcodes in the table')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T20:03:38.954Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Define the template for alignment - should be a fasta file with only one sequence in it \n",
    "TEMPLATE_FILE = 'rbcL_codonopt_barcoded.txt'\n",
    "templateParsed = SeqIO.parse(TEMPLATE_FILE, 'fasta')\n",
    "TEMPLATE_STRING = None\n",
    "for record in templateParsed:\n",
    "    assert TEMPLATE_STRING == None\n",
    "    TEMPLATE_STRING = str(record.seq)\n",
    "    RCtemplateString = record.seq.reverse_complement()\n",
    "    TEMPLATE_TRANSLATION = str(record.seq.translate(to_stop=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T17:14:27.830107Z",
     "start_time": "2021-05-19T17:06:59.856371Z"
    },
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0 percent done\n",
      "0.17 percent done\n",
      "0.34 percent done\n",
      "1.71 percent done\n",
      "3.43 percent done\n",
      "17.1 percent done\n",
      "34.3 percent done\n",
      "68.6 percent done\n"
     ]
    }
   ],
   "source": [
    "# Generate dataframe of unique barcodes, perform consensus functions, save as a .pkl file\n",
    "uniqueBCDF = getConsensus(uniqueBCs)\n",
    "uniqueBCDF = pd.merge(uniqueBCDF, readsDF[['Barcode_sequence', 'BCfrequency']], \n",
    "                      how='left', on = 'Barcode_sequence')\n",
    "uniqueBCDF.drop_duplicates(subset = ['Barcode_sequence'], inplace = True)\n",
    "uniqueBCDF.to_pickle(FILE_NAME + '_consensusTable.pkl')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T19:49:41.236574Z",
     "start_time": "2021-05-19T19:49:38.725059Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Open the saved .pkl file of consensus mutations\n",
    "uniqueBCDF = pd.read_pickle(FILE_NAME + '_consensusTable.pkl')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-19T17:25:36.299431Z",
     "start_time": "2021-05-19T17:22:31.298148Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Apply all filters and save final lookup tables \n",
    "# Reorder so that it goes from more to less common barcodes so that it removes the less common ones\n",
    "uniqueBCDF = uniqueBCDF.sort_values(by=['BCfrequency'], ascending = True).reset_index()\n",
    "# Off-by-one check function\n",
    "\n",
    "# Make a list and then a set to look through, this is important (hash reasons) \n",
    "uniqueBCs = uniqueBCDF['Barcode_sequence'].tolist()\n",
    "uniqueBCLookup = set(uniqueBCs)\n",
    "uniqueBCDF['OBO_found'], uniqueBCDF['Similar_barcode'] = zip(*uniqueBCDF.apply(OBOcheck, axis = 1))\n",
    "\n",
    "# Apply filters (edit this to run through dataframe once somehow)\n",
    "uniqueBCDF['No_Mutations'], uniqueBCDF['No_Consensus'], uniqueBCDF['Indels'] = \\\n",
    "        zip(*uniqueBCDF.apply(mutFilters, axis = 1))\n",
    "uniqueBCDF['No_BB_Mutations'] = uniqueBCDF.apply(lambda row: BBmutFilters(row), axis = 1)\n",
    "uniqueBCDF['Mut_positions'] = uniqueBCDF.apply(lambda row: getMutPos(row), axis = 1)\n",
    "uniqueBCDF['AffectedCodons'] = uniqueBCDF.apply(lambda row: oneCodonOnlyCheck(row.Mut_positions), axis = 1)\n",
    "uniqueBCDF['HowManyAffectedCodons'] = uniqueBCDF.apply(lambda row: len(row.AffectedCodons)\n",
    "                                                            if row.AffectedCodons[0]>-1 else (-1), axis = 1)\n",
    "uniqueBCDF['AffectedCodonInt'] = uniqueBCDF.apply(lambda row: row.AffectedCodons[0]\n",
    "                                                    if (len(row.AffectedCodons)) == 1 else (-2), axis = 1)\n",
    "uniqueBCDF['Original_codon'], uniqueBCDF['Mutant_codon'] = zip(*uniqueBCDF.apply(getCodonSeqs, axis = 1))\n",
    "uniqueBCDF['Codon_Allowed'] = uniqueBCDF.apply(lambda row: \n",
    "                                               True if row.Mutant_codon in PROGRAMMED_MUTATION_CODONS\n",
    "                                               else False, axis = 1)\n",
    "uniqueBCDF['Enough_reads'] = uniqueBCDF.apply(lambda row: True if (row['BCfrequency'] >= MINIMUM_READS)\n",
    "                                               else False, axis = 1)\n",
    "uniqueBCDF['AAMutation'], uniqueBCDF['mutantAA'] = zip(*uniqueBCDF.apply(getAAmut, axis = 1))\n",
    "uniqueBCDF['Passed_all_filters'] = uniqueBCDF.apply(lambda row: passedAllFilters(row), axis=1)\n",
    "\n",
    "# Shift aminoacid numbers by one\n",
    "uniqueBCDF['correctedAAmut'] = uniqueBCDF.apply(lambda row: shiftAAone(row), axis=1)\n",
    "\n",
    "# Count barcodes per mutant\n",
    "uniqueBCDF['BCsForMut'] = uniqueBCDF['correctedAAmut'].map(uniqueBCDF['correctedAAmut'].value_counts())\n",
    "\n",
    "# Save lookup table\n",
    "uniqueBCDF[uniqueBCDF['Passed_all_filters'] == True][['Barcode_sequence', \n",
    "                                                'correctedAAmut']].to_csv(FILE_NAME + '_lookupTable.csv')\n",
    "\n",
    "# Save lookup tables for reads with indels and backbone mutations, filter with read cutoff\n",
    "uniqueBCDF[(uniqueBCDF['Enough_reads'] == True) & \n",
    "           (uniqueBCDF['No_BB_Mutations'] == False)][['Barcode_sequence', \n",
    "            'correctedAAmut']].to_csv(FILE_NAME + '_indels_lookupTable.csv')\n",
    "uniqueBCDF[(uniqueBCDF['Enough_reads'] == True) & \n",
    "           (uniqueBCDF['Indels'] == True)][['Barcode_sequence', \n",
    "            'correctedAAmut']].to_csv(FILE_NAME + '_BBmuts_lookupTable.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:45.423Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Add columns to reads dataframe\n",
    "\n",
    "readsDF = pd.merge(readsDF, uniqueBCDF[['Barcode_sequence', \\\n",
    "                                        'Consensus_mutations', \\\n",
    "                                        'OBO_found', \\\n",
    "                                        'No_Mutations', \\\n",
    "                                        'No_Consensus', \\\n",
    "                                        'Indels', \\\n",
    "                                        'HowManyAffectedCodons', \\\n",
    "                                        'Codon_Allowed', \\\n",
    "                                        'AAMutation', \\\n",
    "                                        'mutantAA', \\\n",
    "                                        'Enough_reads', \\\n",
    "                                        'No_BB_Mutations', \\\n",
    "                                        'Passed_all_filters']], \\\n",
    "                                        how='left', on = 'Barcode_sequence')\n",
    "readsDF.head(2)\n",
    "\n",
    "filteredReadDFs = filterReads(readsDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:45.838Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Rarefaction curves data\n",
    "\n",
    "rarefactionData = {}\n",
    "\n",
    "for label in filterLabels:\n",
    "    rarefactionData[label] = []\n",
    "\n",
    "n = [10, 30, 100, 300, 1000, 3000, 10000, 30000]\n",
    "howOftenToSample = 100000\n",
    "\n",
    "readsDF['index'] = np.arange(len(readsDF))\n",
    "\n",
    "for read in readsDF['index']:\n",
    "    num=read%howOftenToSample\n",
    "    if (num == 0) or (read in n):\n",
    "        for label in filterLabels:\n",
    "            filteredBCsUpToRead = filteredReadDFs[label]['Barcode_sequence'][:read]\n",
    "            rarefactionData[label].append([filteredBCsUpToRead.count(), filteredBCsUpToRead.nunique()])\n",
    "            \n",
    "rarefactionDataTransposed = {}\n",
    "# Transpose values\n",
    "for label in filterLabels:\n",
    "    BCcounts, unique_BCs = list(map(list, zip(*rarefactionData[label])))\n",
    "    rarefactionDataTransposed[label] = BCcounts, unique_BCs        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:46.334Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Plot rarefaction curves\n",
    "for label in filterLabels:\n",
    "    plt.scatter(rarefactionDataTransposed[label][0], rarefactionDataTransposed[label][1])\n",
    "\n",
    "plt.legend(filterLabels, bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "plt.title('Barcode subsampling')\n",
    "plt.xlabel('Barcodes in subsample')\n",
    "plt.ylabel('Unique barcodes in subsample')\n",
    "\n",
    "# Fit to negative binomial distribution curve\n",
    "\n",
    "from scipy.optimize import curve_fit\n",
    "\n",
    "def negativeBinomial(x, a, b):\n",
    "    return a*(1-np.power(b, x/a))  # a*(1-b^(x/a))\n",
    "\n",
    "xdata = rarefactionDataTransposed['Fully filtered'][0]\n",
    "ydata = rarefactionDataTransposed['Fully filtered'][1]\n",
    "\n",
    "popt, pcov = curve_fit(f = negativeBinomial, \\\n",
    "                      xdata = xdata, \\\n",
    "                      ydata = ydata, \\\n",
    "                      p0=[10000, 0.5], \\\n",
    "                      bounds=(0, np.inf))\n",
    "\n",
    "estText = \"Negative binomial \\ncomplexity estimate: \\n\" + str(int(popt[0])) + ' barcodes'\n",
    "print(\"Negative binomial skew estimate: \" + str(popt[1])[:4])\n",
    "\n",
    "# Plot the curve above\n",
    "plt.plot(xdata, negativeBinomial(xdata, *popt), color = 'black')\n",
    "plt.tight_layout()\n",
    "plt.rcParams[\"figure.figsize\"] = (7,7)\n",
    "plt.text(2500000, 50000, estText, fontsize=12, horizontalalignment='center', verticalalignment='bottom')\n",
    "\n",
    "plt.savefig(FILE_NAME + '_barcodeSubsamplingByFilter.pdf')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:46.828Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Make histograms for the same data\n",
    "\n",
    "binSet = range(0,100,1)\n",
    "counts = []\n",
    "bins = []\n",
    "bars = []\n",
    "\n",
    "for i, aFilter in enumerate(filterLabels):\n",
    "    readHistData = filteredReadDFs[aFilter]['BCfrequency']\n",
    "    histtype = 'step'\n",
    "    if aFilter == 'Fully filtered':\n",
    "        histtype = 'stepfilled'\n",
    "    x = plt.hist(readHistData, alpha = 0.3, bins = binSet, histtype = histtype, label = aFilter)\n",
    "    counts.append(x[0])\n",
    "    bins.append(x[1])\n",
    "    bars.append(x[2])\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (7,7)\n",
    "plt.legend()\n",
    "plt.title('Reads per barcode')\n",
    "plt.ylabel('Barcodes')\n",
    "plt.xlabel('Reads per barcode')\n",
    "plt.savefig(FILE_NAME + '_preliminary_readsHist.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:47.235Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Downsampled data histograms\n",
    "m = [50000, 100000, 200000, 500000, 1000000]\n",
    "\n",
    "BCList = filteredReadDFs['Fully filtered']['Barcode_sequence']\n",
    "\n",
    "for i, barcode in enumerate(BCList):\n",
    "    if i in m:\n",
    "        plt.hist(filteredReadDFs['Fully filtered'][:i].groupby('AAMutation')['Barcode_sequence'].nunique(), \\\n",
    "                 bins = range(0,50,1), \\\n",
    "                 histtype = 'step', \\\n",
    "                 alpha = 0.7, density = True, label = 'First ' + str(i) + ' reads')\n",
    "    if i == len(BCList) - 1:\n",
    "        plt.hist(filteredReadDFs['Fully filtered'][:i].groupby('AAMutation')['Barcode_sequence'].nunique(), \\\n",
    "                 bins = range(0,50,1), \\\n",
    "                 histtype = 'step', \\\n",
    "                 edgecolor = 'black', density = True, label = 'All filtered reads (' + str(len(BCList)) + ')')\n",
    "plt.legend()\n",
    "plt.title('Barcodes per mutation subsamples')\n",
    "plt.ylabel('Mutations')\n",
    "plt.xlabel('Barcodes per Mutation')\n",
    "\n",
    "plt.savefig(FILE_NAME + '_barcodesPerMutHistSubsampled.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:47.471Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Histogram of barcodes per mutation\n",
    "\n",
    "BCsPerMut = filteredReadDFs['Fully filtered'].groupby('AAMutation')['Barcode_sequence'].nunique()\n",
    "missedMuts = TOTAL_MUTS - len(BCsPerMut)\n",
    "BCsPerMut = BCsPerMut.append(pd.Series([0]*missedMuts))\n",
    "\n",
    "binSet = range(0,50,1)\n",
    "\n",
    "histData = plt.hist(BCsPerMut, bins = binSet, color = 'lightgray')\n",
    "plt.title('Barcodes per mutation')\n",
    "plt.ylabel('Mutations')\n",
    "plt.xlabel('Barcodes per Mutation')\n",
    "\n",
    "plt.savefig(FILE_NAME + '_barcodesPerMutHist.pdf')\n",
    "\n",
    "# Fit poisson or gamma curve\n",
    "# Print out median or whatever\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:47.944Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Make a graph showing how many mutants I have at each position of the protein\n",
    "# Ideally I would have 19 mutants at each position\n",
    "\n",
    "def getMutPosition(mutation):\n",
    "    return int(mutation[1:-1])\n",
    "\n",
    "def getMutAA(mutation):\n",
    "    return mutation[-1:]\n",
    "\n",
    "foundMutPos = []\n",
    "foundMutAA = []\n",
    "\n",
    "for mutant in filteredReadDFs['Fully filtered']['AAMutation']:\n",
    "    foundMutPos.append(getMutPosition(mutant))\n",
    "    foundMutAA.append(getMutAA(mutant))\n",
    "        \n",
    "mutationPositionData = {'Position' : foundMutPos , 'Amino acid' : foundMutAA}\n",
    "dfPositionAA = pd.DataFrame(mutationPositionData)\n",
    "mutationsByPosition = dfPositionAA.groupby('Position')['Amino acid'].nunique()\n",
    "dfPositionAAanalyzed = pd.DataFrame(mutationsByPosition)\n",
    "dfPositionAAanalyzed.to_csv(FILE_NAME + '_mutationsByPosition.csv')\n",
    "mutationBCsByPosition = dfPositionAA.groupby('Position')['Amino acid'].count()\n",
    "\n",
    "fig,ax = plt.subplots()\n",
    "plt.gcf().set_size_inches((20, 5))    \n",
    "\n",
    "ax.scatter(x = dfPositionAAanalyzed.index, \\\n",
    "            y = dfPositionAAanalyzed['Amino acid'], \\\n",
    "            s = 5, \\\n",
    "            color = 'black', \\\n",
    "            marker = \"_\")\n",
    "ax.set_xlabel(\"Amino acid position\",fontsize=14)\n",
    "\n",
    "ax.set_ylabel(\"Mutantions at that position\",fontsize=14)\n",
    "ax.set_ylim([0, 20])\n",
    "ax.set_xlim([-1, 465])\n",
    "ax.set_yticks([0, 2, 5, 10, 15, 18, 19])\n",
    "\n",
    "# twin object for two different y-axis on the sample plot\n",
    "ax2=ax.twinx()\n",
    "# make a plot with different y-axis using second axis object\n",
    "ax2.scatter(mutationBCsByPosition.index, mutationBCsByPosition, s = 10, color = 'green')\n",
    "ax2.set_ylabel(\"Barcodes of mutations at that position\",fontsize=14)\n",
    "\n",
    "plt.savefig(FILE_NAME + '_coverageByPosition.pdf')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:48.316Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Calculate mutant coverage\n",
    "totalAALength = 466\n",
    "DMSAALength = 462\n",
    "totalPossibleMuts = 19* totalAALength\n",
    "DMSMuts = 19 * DMSAALength\n",
    "\n",
    "mutationsAfterFilters = mutationsByPosition.sum()\n",
    "\n",
    "# Calculate percent of possible mutations that we have\n",
    "percentMuts = mutationsAfterFilters/totalPossibleMuts\n",
    "percentDMSMuts = mutationsAfterFilters/DMSMuts\n",
    "mutsW3BCsorMore = (BCsPerMut > 2).value_counts()[True]\n",
    "\n",
    "\n",
    "print('We captured ' + str(100*percentMuts)[:4] + '% of possible single mutants (' + \\\n",
    "      str(mutationsAfterFilters) + ' of ' + str(totalPossibleMuts) + ')')\n",
    "print('We captured ' + str(100*percentDMSMuts)[:4] + '% of possible single mutants we ordered (' + \\\n",
    "      str(mutationsAfterFilters) + ' of ' + str(DMSMuts) + ')')\n",
    "print('We captured ' + str(100*mutsW3BCsorMore/DMSMuts)[:4] + '% of possible single mutants we ordered (' + \\\n",
    "      str(mutsW3BCsorMore) + ' of ' + str(DMSMuts) + ') with 3 barcodes or more')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:48.698Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Heatmap showing each individual mutation barcode frequency\n",
    "# This will only give the barcode heatmap for now\n",
    "# The positions here are indexed to 0 so they're sort of off by 3\n",
    "\n",
    "positionAAFreq = uniqueBCDF[uniqueBCDF['Passed_all_filters'] == True].groupby(['AffectedCodonInt', \n",
    "                                                                               'mutantAA']).size()\n",
    "dfPositionAAFreq = positionAAFreq.to_frame(name = 'Barcodes')\n",
    "pivotBCs = dfPositionAAFreq.pivot_table(index = 'mutantAA', columns = 'AffectedCodonInt', values = 'Barcodes')\n",
    "\n",
    "fig, (ax1) = plt.subplots(1, tight_layout=True, figsize=(10,5), dpi = 100)\n",
    "pcmBCs = ax1.pcolor(pivotBCs, cmap = 'viridis', norm=matplotlib.colors.LogNorm())\n",
    "\n",
    "ax1.set_title('Barcodes per mutant', size = 15)\n",
    "ax1.set_xlabel('Amino acid position in rbcL', size = 15)\n",
    "ax1.set_ylabel('Amino acid', size = 15)\n",
    "# This sets the ticks in the middle\n",
    "ax1.set_yticks(np.arange(0.5, 20.5))\n",
    "ax1.set_yticklabels(pivotBCs.index)\n",
    "fig.colorbar(pcmBCs, ax = ax1)\n",
    "\n",
    "plt.savefig(FILE_NAME + '_mutationBCHist.pdf')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-19T17:25:49.170Z"
    },
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Run cumulative version of filters to make pie chart\n",
    "\n",
    "startingBCs = len(filteredReadDFs['All data'])\n",
    "\n",
    "cumulativeDF = filteredReadDFs['All data']\n",
    "\n",
    "filteredReadsList = []\n",
    "\n",
    "for i, aFilter in enumerate(filteredReadDFs.keys()):\n",
    "    readsRemaining = len(cumulativeDF)\n",
    "    cumulativeDF = cumulativeDF[['query_name', \n",
    "                'Barcode_sequence']].merge(filteredReadDFs[aFilter][['query_name']], \n",
    "                                           on = 'query_name', how = 'inner')\n",
    "    filteredReads = readsRemaining - len(cumulativeDF)\n",
    "    filteredReadsList.append(filteredReads)\n",
    "#     print(aFilter)\n",
    "#     print('This filter removed ' + str(filteredReads) + ' reads')\n",
    "#     print(cumulativeDF['Barcode_sequence'].count())\n",
    "#     print(cumulativeDF['Barcode_sequence'].nunique())\n",
    "\n",
    "# This gives the remaining reads as the final slice\n",
    "filteredReadsList[-1] = readsRemaining\n",
    "    \n",
    "readSummary = {filterLabels[i]: filteredReadsList[i] for i in range(len(filterLabels))} \n",
    "print(readSummary)\n",
    "    \n",
    "dfReadSummary = pd.DataFrame(readSummary.items(), columns=['Read type', FILE_NAME])\n",
    "dfReadSummary.to_csv(FILE_NAME + '_readSummary.csv')\n",
    "\n",
    "pieChart = dfReadSummary.plot.pie(y = FILE_NAME, \\\n",
    "                             labels = dfReadSummary['Read type'], \\\n",
    "                             figsize=(7, 7), \\\n",
    "                             autopct='%1.1f%%', \\\n",
    "                             cmap = 'Set3_r', \\\n",
    "                             startangle = 0)\n",
    "\n",
    "pieChart.get_legend().remove()\n",
    "\n",
    "plt.savefig(FILE_NAME + '_readPieChart.pdf')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
